diff --git a/megatron/core/optimizer/distrib_optimizer.py b/megatron/core/optimizer/distrib_optimizer.py
index 2925edcce..19f557a3b 100644
--- a/megatron/core/optimizer/distrib_optimizer.py
+++ b/megatron/core/optimizer/distrib_optimizer.py
@@ -33,6 +33,7 @@ except ImportError:
         HAVE_APEX_OR_TE = False
 
 from megatron.core.optimizer.cpu_offloading import HybridDeviceOptimizer
+from megatron.core.parallel_state import get_tensor_model_parallel_group
 
 from .. import tensor_parallel
 from ..config_logger import has_config_logger_enabled, log_config_to_disk
@@ -51,6 +52,7 @@ from ..transformer.module import MegatronModule
 from .grad_scaler import MegatronGradScaler
 from .optimizer import MixedPrecisionOptimizer, _zero_grad_group_helper, param_group_identifier_keys
 from .optimizer_config import OptimizerConfig
+from .muon import Muon, MuonDistMeta
 
 logger = getLogger(__name__)
 
@@ -158,6 +160,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                 sub_param_start = max(0, gbuf_world_range.start - param_world_start)
                 sub_param_range = param_local_range.normalize(sub_param_start)
                 param_range_map[param] = {
+                    "world_indexes": (param_world_start, param_world_end),
                     "gbuf_world": param_world_range,
                     "gbuf_world_in_bucket": param_world_range_in_bucket,
                     "gbuf_local": param_local_range,
@@ -345,13 +348,24 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
             shard_fp32_groups.append(shard_fp32_params_this_group)
             shard_fp32_from_float16_groups.append(shard_fp32_from_float16_params_this_group)
 
+            dist_metas = {}
             for model_param in group_range["params"]:
 
                 assert model_param.requires_grad
 
                 gbuf_index, dtype, bucket_index = param_gbuf_map[model_param]
                 gbuf_range = gbuf_ranges[gbuf_index][dtype][bucket_index]
-                param_range = gbuf_range["param_map"][model_param]["param"]
+                param_gbuf_ranges = gbuf_range["param_map"][model_param]
+                param_range = param_gbuf_ranges["param"]
+
+                if config.optimizer == "muon":
+                    # gen dist meta
+                    param_world_indexes = param_gbuf_ranges["world_indexes"]
+                    # To get the tensor parallel split dim
+                    tp_split_dim = getattr(model_param, 'partition_dim') if getattr(
+                        model_param, 'tensor_model_parallel', False) else -1
+                    dist_meta = MuonDistMeta(
+                        gbuf_index, bucket_index, model_param.shape, param_world_indexes, tp_split_dim)
 
                 # fp16, bf16 params.
                 if model_param.type() in ['torch.cuda.HalfTensor', 'torch.cuda.BFloat16Tensor']:
@@ -386,7 +400,9 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                                     .to(model_param.device)
                                     .float()
                                 )
-                                model_param.clear_high_precision_init_val()
+                                # When --fp8-param-gather on, the high precision value
+                                # should leave for further initializing the main param.
+                                # model_param.clear_high_precision_init_val()
                             else:
                                 shard_main_param = model_param.float().view(-1)[
                                     param_range.start : param_range.end
@@ -412,6 +428,13 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                     shard_float16_params_this_group.append(shard_model_param)
                     shard_fp32_from_float16_params_this_group.append(shard_main_param)
 
+                    # add to dist metas
+                    if config.optimizer == "muon":
+                        if config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
+                            dist_metas[shard_model_param] = dist_meta
+                        else:
+                            dist_metas[shard_main_param] = dist_meta
+
                 # fp32 params.
                 elif model_param.type() == 'torch.cuda.FloatTensor':
                     shard_model_param = model_param.view(-1)[param_range.start : param_range.end]
@@ -450,7 +473,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
             shard_float16_groups,
             shard_fp32_groups,
             shard_fp32_from_float16_groups,
-        )
+        ), dist_metas
 
     def __init__(
         self,
@@ -508,10 +531,10 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
         self.distributed_optimizer_instance_id = distributed_optimizer_instance_id
 
         assert (
-            isinstance(optimizer, (Adam, torch.optim.AdamW, HybridDeviceOptimizer))
+            isinstance(optimizer, (Adam, torch.optim.AdamW, HybridDeviceOptimizer, Muon))
             or optimizer is None
         ), (
-            "Only Adam and HybridDeviceOptimizer currently supported, "
+            "Only Adam / HybridDeviceOptimizer / Muon currently supported, "
             "due to checkpointing requirements."
         )
 
@@ -591,7 +614,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
             self.shard_float16_groups,
             self.shard_fp32_groups,
             self.shard_fp32_from_float16_groups,
-        ) = self._build_model_and_main_param_groups(
+        ), dist_metas = self._build_model_and_main_param_groups(
             self.gbuf_ranges, self.model_param_gbuf_map, self.opt_group_ranges, config
         )
 
@@ -599,10 +622,32 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
             self.optimizer = HybridDeviceOptimizer(
                 params=[g["orig_group"] for g in self.opt_group_ranges], **self.optimizer.defaults
             )
+            if config.optimizer == 'muon':
+                assert all(grad_buffer.grad_dtype == torch.float32 for grad_buffer in self.buffers), \
+                    "all grad buffer should only contains float32 type for muon optimizer"
+                gbuf_sizes = [ [(bucket.grad_data.numel(), bucket.offset) for bucket in buffer.buckets ]
+                                for buffer in self.buffers ]
+                self.optimizer.cpu_optimizers[0].enable_distributed_mode(
+                    gbuf_sizes, self.data_parallel_group,
+                    get_tensor_model_parallel_group(),
+                    dist_metas,
+                )
         else:
             self.optimizer.param_groups = [g["orig_group"] for g in self.opt_group_ranges]
             self.optimizer.load_state_dict(self.optimizer.state_dict())
 
+            # for muon optimizer, enable distributed mode
+            if isinstance(self.optimizer, Muon):
+                assert all(grad_buffer.grad_dtype == torch.float32 for grad_buffer in self.buffers), \
+                    "all grad buffer should only contains float32 type for muon optimizer"
+                gbuf_sizes = [[(bucket.grad_data.numel(), bucket.offset) for bucket in buffer.buckets]
+                                for buffer in self.buffers ]
+                self.optimizer.enable_distributed_mode(
+                    gbuf_sizes, self.data_parallel_group,
+                    get_tensor_model_parallel_group(),
+                    dist_metas,
+                )
+
     def _get_model_param_range_map(self, param: torch.nn.Parameter):
         """
         Given a model param, get the index sub-range of the param that this
@@ -647,6 +692,13 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                     assert len(steps) == 1, f"steps: {optimizer.state}"
                     step = steps[0]
                     break
+                elif self.config.use_deepspeed_cpu_adam:
+                    if len(optimizer.state) == 0:
+                        continue
+                    steps = list(set([s["step"] for s in optimizer.state.values()]))
+                    assert len(steps) == 1, f"steps: {optimizer.state}"
+                    step = steps[0]
+                    break
         elif USING_TE_OPTIMIZER or USING_APEX_OPTIMIZER:
             # Extract 'step', for TE FusedAdam support.
             steps = list(
@@ -784,8 +836,8 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                             # Allocate dummy tensors.
                             numel = len(param_range_map["gbuf_world"])
                             init_shard = lambda dtype=torch.float32: torch.empty(
-                                (numel,), dtype=dtype, device=torch.cuda.current_device()
-                            )
+                                (numel,), dtype=dtype, device="cpu"
+                            )  # initialize the dummy tensor on CPU side to avoid GPU mem overhead
 
                             # For precision_aware_optimizer, the empty tensors should also be
                             #  initialized with the correct dtype.
@@ -793,6 +845,15 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                                 "exp_avg": init_shard(self.config.exp_avg_dtype),
                                 "exp_avg_sq": init_shard(self.config.exp_avg_sq_dtype),
                             }
+                            # for muon optimizer link state ( for load state dict )
+                            if self.config.optimizer == "muon":
+                                use_muon = self.optimizer.param_groups[group_index].get("use_muon", False)
+                                if use_muon:
+                                    tensors["muon_buffer"] = tensors.pop("exp_avg")
+                                    tensors.pop("exp_avg_sq", None)
+                                else:
+                                    tensors["adamw_exp_avg"] = tensors.pop("exp_avg")
+                                    tensors["adamw_exp_avg_sq"] = tensors.pop("exp_avg_sq")
                             if self.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
                                 if self.config.store_param_remainders and self.config.bf16:
                                     tensors["master_param"] = init_shard(torch.int16)
@@ -894,6 +955,17 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
             main_param = self.optimizer.param_groups[group_index]["params"][group_order]
             optim_state = self.optimizer.state[main_param]
             tensors = {"param": main_param, **optim_state}
+
+        # process muon to be compatiable with adam ( always save to exp_avg / exp_avg_sq )
+        if isinstance(self.optimizer, Muon):
+            use_muon = self.optimizer.param_groups[group_index].get("use_muon", False)
+            if use_muon:
+                tensors["exp_avg"] = tensors["muon_buffer"]
+                tensors["exp_avg_sq"] = torch.zeros_like(tensors["param"])
+            else:
+                tensors["exp_avg"] = tensors["adamw_exp_avg"]
+                tensors["exp_avg_sq"] = tensors["adamw_exp_avg_sq"]
+
         return tensors
 
     def _set_main_param_and_optimizer_states(self, model_param, tensors):
@@ -907,25 +979,50 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
         }
         """
         group_index, group_order = self.model_param_group_index_map[model_param]
+
+        # for muon optimizer link state ( for load state dict )
+        if self.config.optimizer == "muon":
+            use_muon = self.optimizer.param_groups[group_index].get("use_muon", False)
+            if use_muon:
+                tensors["muon_buffer"] = tensors.pop("exp_avg")
+                tensors.pop("exp_avg_sq", None)
+            else:
+                tensors["adamw_exp_avg"] = tensors.pop("exp_avg")
+                tensors["adamw_exp_avg_sq"] = tensors.pop("exp_avg_sq")
+
         if self.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
             sharded_model_param = self.optimizer.param_groups[group_index]["params"][group_order]
             for k, v in tensors.items():
                 if isinstance(self.optimizer, HybridDeviceOptimizer):
                     if k == "param":
                         k = "master_param"
+                        sharded_model_param.copy_(v)
                     self.optimizer.state[sharded_model_param][k] = v
                     continue
 
                 if k == "param":
+                    if self.config.store_param_remainders and self.config.bf16:
+                        v = v.to(torch.int16)
                     self.optimizer.set_scaled_state(sharded_model_param, "master_param", v)
                 else:
+                    if v.dtype != torch.float32:
+                        v = v.to(torch.float32)
                     self.optimizer.set_scaled_state(sharded_model_param, k, v)
         else:
             main_param = self.optimizer.param_groups[group_index]["params"][group_order]
             optim_state = self.optimizer.state[main_param]
-            dst_tensors = {"param": main_param, **optim_state}
-            for key in dst_tensors:
-                dst_tensors[key].copy_(tensors[key])
+            if isinstance(self.optimizer, HybridDeviceOptimizer):
+                for k, v in tensors.items():
+                    if k == "param":
+                        k = "master_param"
+                    optim_state[k] = v
+            else:
+                dst_tensors = {"param": main_param, **optim_state}
+                for key in dst_tensors:
+                    # since muon optimizer does not have exp_avg and exp_avg_sq (we pop them before)
+                    if self.config.optimizer == "muon" and not key in tensors:
+                        continue
+                    dst_tensors[key].copy_(tensors[key])
 
     def get_parameter_state_dp_reshardable(self):
         """Get internal representation of parameter state without any copies and modifications.
@@ -962,6 +1059,23 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
             state[gbuf_idx] = dtype_state
         return state
 
+    def _get_dtype_by_key(self, key):
+        """Get data type by key.
+        Args:
+            key (str): Key of the optimizer state.
+
+        Returns:
+            torch.dtype: Data type of the optimizer state.
+        """
+        if key == "param":
+            return torch.float32
+        elif key == "exp_avg":
+            return self.config.exp_avg_dtype
+        elif key == "exp_avg_sq":
+            return self.config.exp_avg_sq_dtype
+        else:
+            raise ValueError(f"Invalid key: {key}")
+    
     def get_parameter_state_dp_zero(
         self,
         use_gloo_comm: bool = True,
@@ -1018,7 +1132,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                 if data_parallel_rank == 0 or return_on_all_ranks:
                     world_tensors = {
                         key: torch.zeros(
-                            (buffer_numel_unpadded,), dtype=torch.float32, device="cpu"
+                            (buffer_numel_unpadded,), dtype=self._get_dtype_by_key(key), device="cpu"
                         )
                         for key in ("param", "exp_avg", "exp_avg_sq")
                     }
@@ -1041,7 +1155,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                         assert gbuf_world_numel_unpadded <= gbuf_world_numel
 
                         local_shards = {
-                            key: torch.zeros((gbuf_local_numel,), dtype=torch.float32, device="cpu")
+                            key: torch.zeros((gbuf_local_numel,), dtype=self._get_dtype_by_key(key), device="cpu")
                             for key in ("param", "exp_avg", "exp_avg_sq")
                         }
 
@@ -1065,7 +1179,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                                 device = "cpu" if use_gloo_comm else torch.cuda.current_device()
                                 recv_tensors = [
                                     torch.zeros(
-                                        (gbuf_local_numel,), dtype=torch.float32, device=device
+                                        (gbuf_local_numel,), dtype=self._get_dtype_by_key(key), device=device
                                     )
                                     for _ in range(data_parallel_world_size)
                                 ]
@@ -1882,7 +1996,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
 
                         # Contiguous local shards (received from DP rank 0).
                         recv_tensor = torch.zeros(
-                            (gbuf_local_numel,), dtype=torch.float32, device="cpu"
+                            (gbuf_local_numel,), dtype=self._get_dtype_by_key(key), device="cpu"
                         )
 
                         # Scatter tensor list.
@@ -1996,7 +2110,7 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
 
                         # Contiguous local shards (received from DP rank 0).
                         recv_tensor = torch.zeros(
-                            (gbuf_local_numel,), dtype=torch.float32, device="cpu"
+                            (gbuf_local_numel,), dtype=self._get_dtype_by_key(key), device="cpu"
                         )
 
                         # Scatter tensor list.
@@ -2043,6 +2157,9 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                 for model_param, tensors in recv_tensors.items():
                     self._set_main_param_and_optimizer_states(model_param, tensors)
 
+        if isinstance(self.optimizer, HybridDeviceOptimizer):
+            self.optimizer._sync_hdo_state_to_sub_optimizers()
+
     @torch.no_grad()
     def load_parameter_state_from_fully_reshardable(self, state_dict: dict):
         """Load counterpart of sharded_param_state_fully_reshardable.
@@ -2523,17 +2640,6 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
         the model params. This copy does not make use of the grad buffer as
         an intermediary.
         """
-        if isinstance(self.optimizer, HybridDeviceOptimizer):
-            self.optimizer.update_fp32_param_by_new_param()
-            return
-
-        if self.ddp_config.use_megatron_fsdp:
-            return
-
-        # When using precision-aware optimizer, main params are held by self.optimizer. It will also
-        # do the work of copying data from main params to model params.
-        if self.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
-            return
 
         if state_dict is not None:
             # Build a mapping from the model params to the corresponding tensors in the state dict,
@@ -2558,15 +2664,45 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
                         model_param = model_param_to_state_dict_param_map[model_param]
 
                     if is_float8tensor(model_param):
-                        shard_model_param = dequantize_fp8_tensor(model_param).view(-1)[
-                            param_range.start : param_range.end
-                        ]
+                        if hasattr(model_param, 'get_high_precision_init_val'):
+                            shard_model_param = (
+                                model_param.get_high_precision_init_val()
+                                .view(-1)[param_range.start : param_range.end]
+                                .clone()
+                                .to(model_param.device)
+                                .float()
+                            )
+                            model_param.clear_high_precision_init_val()
+                        else:
+                            shard_model_param = dequantize_fp8_tensor(model_param).view(-1)[
+                                param_range.start : param_range.end
+                            ]
                     else:
                         shard_model_param = model_param.view(-1)[
                             param_range.start : param_range.end
                         ]
                     shard_main_param.data.copy_(shard_model_param)
 
+        if isinstance(self.optimizer, HybridDeviceOptimizer):
+            # Copy model groups to shard groups. 
+            # HDO uses `shard_params` as param_groups:
+            # bf16 case, `shard_params` shares the same underlying storage, after loading ckpt, it was updated.
+            # fp8 case, `shard_params` is fp32 copy of model params, after loading ckpt, it is not updated, so the 
+            # explicit "model param -> shard param copy" is needed.
+            if not self.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
+                copy_group_params(self.model_float16_groups, self.shard_fp32_from_float16_groups)
+                copy_group_params(self.model_fp32_groups, self.shard_fp32_groups)
+            self.optimizer.update_fp32_param_by_new_param()
+            return
+
+        if self.ddp_config.use_megatron_fsdp:
+            return
+
+        # When using precision-aware optimizer, main params are held by self.optimizer. It will also
+        # do the work of copying data from main params to model params.
+        if self.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
+            return
+
         # Copy model groups to shard groups.
         copy_group_params(self.model_float16_groups, self.shard_fp32_from_float16_groups)
         copy_group_params(self.model_fp32_groups, self.shard_fp32_groups)
