diff --git a/megatron/core/optimizer/optimizer_config.py b/megatron/core/optimizer/optimizer_config.py
index 8151d5e9d..15b66e387 100644
--- a/megatron/core/optimizer/optimizer_config.py
+++ b/megatron/core/optimizer/optimizer_config.py
@@ -43,6 +43,11 @@ class OptimizerConfig:
     fp8_recipe: Optional[str] = None
     """The type of fp8 recipe will affect the processing logic inside distributed optimizer."""
 
+    fp8_param_gather: bool = False
+    """It is the same as fp8_param in TransformerConfig and fp8_param_gather in DistributedDataParallelConfig.
+       If true, keep the compute param in fp8 (do not use any other intermediate dtype) and
+       perform the param all-gather in fp8."""
+
     fp16: bool = False
     """If true, train with fp16 mixed precision training. Defaults to False."""
 
@@ -124,6 +129,19 @@ class OptimizerConfig:
     sgd_momentum: float = 0.9
     """Momentum factor for SGD optimizer."""
 
+    # Muon.
+    muon_momentum: float = 0.95
+    """Momentum factor for Muon optimizer."""
+    
+    muon_nesterov: bool = True
+    """If true, use Nesterov momentum in Muon optimizer."""
+
+    muon_ns_steps: int = 5
+    """Number of Newton-Schulz iteration steps"""
+
+    muon_matched_adamw_rms: float = 0.2
+    """The adamw update rms that muon is designed to matched, typicially 0.2 ~ 0.4"""
+
     #######################
     # Distributed optimizer
     #######################
@@ -168,6 +186,9 @@ class OptimizerConfig:
     pin_cpu_params: bool = True
     """If True, pin the optimizer parameters to CPU memory."""
 
+    use_deepspeed_cpu_adam: bool = True
+    """If True, use DeepSpeed CPU Adam implementation instead of Torch CPU Adam."""
+
     ################
     # Miscellaneous
     ################
@@ -200,7 +221,7 @@ class OptimizerConfig:
             and (
                 self.main_params_dtype != torch.float32
                 or (self.fp8_recipe is None or self.fp8_recipe == "delayed")
-                or self.optimizer_cpu_offload
+                or (self.optimizer_cpu_offload and not self.fp8_param_gather)
             )
         )
 
@@ -217,8 +238,8 @@ class OptimizerConfig:
 
         if self.use_precision_aware_optimizer:
             assert (
-                self.optimizer == 'adam'
-            ), '--use-precision-aware-optimizer only supported with adam'
+                self.optimizer == 'adam' or self.optimizer == 'muon'
+            ), '--use-precision-aware-optimizer supported with adam or muon'
             assert (
                 self.use_distributed_optimizer
             ), '--use-precision-aware-optimizer only supported with distributed optimizer'
