diff --git a/transformer_engine/pytorch/attention/rope.py b/transformer_engine/pytorch/attention/rope.py
index cc23d65a..b9e43b99 100644
--- a/transformer_engine/pytorch/attention/rope.py
+++ b/transformer_engine/pytorch/attention/rope.py
@@ -129,6 +129,7 @@ class FusedRoPEFunc(torch.autograd.Function):
         cu_seqlens: Union[torch.Tensor, None] = None,
         cp_size: int = 1,
         cp_rank: int = 0,
+        mscale: float = 1.0,
     ) -> torch.Tensor:
         """Fused RoPE forward."""
 
@@ -148,12 +149,14 @@ class FusedRoPEFunc(torch.autograd.Function):
             cu_seqlens,
             cp_size,
             cp_rank,
+            mscale,
         )
         ctx.save_for_backward(freqs, cu_seqlens)
         ctx.tensor_format = tensor_format
         ctx.cp_size = cp_size
         ctx.cp_rank = cp_rank
         ctx.interleaved = interleaved
+        ctx.mscale = mscale
 
         return output
 
@@ -169,9 +172,10 @@ class FusedRoPEFunc(torch.autograd.Function):
             cu_seqlens,
             ctx.cp_size,
             ctx.cp_rank,
+            ctx.mscale,
         )
 
-        return grad_input, None, None, None, None, None, None, None
+        return grad_input, None, None, None, None, None, None, None, None
 
 
 class FusedQKVRoPEFunc(torch.autograd.Function):
@@ -379,6 +383,7 @@ def apply_rotary_pos_emb(
     cu_seqlens: Union[torch.Tensor, None] = None,
     cp_size: int = 1,
     cp_rank: int = 0,
+    mscale: float = 1.0,
 ) -> torch.Tensor:
     """
     Apply rotary positional embedding tensor to the input tensor.
@@ -422,6 +427,8 @@ def apply_rotary_pos_emb(
         Context parallel world size. Only valid when `tensor_format` is 'thd' and `fused` is True.
     cp_rank: int, default = 0.
         Context parallel rank. Only valid when `tensor_format` is 'thd' and `fused` is True.
+    mscale: float, default = 1.0
+        The scaling factor for the yarn rotary positional embedding. Only valid when `fused` is True.
     """
 
     # `start_positions` is only supported for `cp_size=1` and inference.
@@ -435,7 +442,7 @@ def apply_rotary_pos_emb(
 
     if fused:
         return FusedRoPEFunc.apply(
-            t, freqs, start_positions, tensor_format, interleaved, cu_seqlens, cp_size, cp_rank
+            t, freqs, start_positions, tensor_format, interleaved, cu_seqlens, cp_size, cp_rank, mscale
         )
 
     # Unfused THD format
