diff --git a/xpu_plugin/multi_latent_attention/__init__.py b/xpu_plugin/multi_latent_attention/__init__.py
new file mode 100644
index 0000000..8957978
--- /dev/null
+++ b/xpu_plugin/multi_latent_attention/__init__.py
@@ -0,0 +1,8 @@
+"""
+Copyright (c) 2026 Baidu.com, Inc. All Rights Reserved
+This module provides megatron-core-aiak plugin.
+"""
+
+from .multi_latent_attention import MLASelfAttention
+
+__all__ = ["MLASelfAttention"]
\ No newline at end of file
