diff --git a/megatron/core/transformer/moe/experts.py b/megatron/core/transformer/moe/experts.py
index d8dd3d03f..1d3b658d9 100644
--- a/megatron/core/transformer/moe/experts.py
+++ b/megatron/core/transformer/moe/experts.py
@@ -1,6 +1,7 @@
 # Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
 
 import copy
+import logging
 import itertools
 from copy import deepcopy
 from functools import partial, wraps
@@ -26,6 +27,11 @@ from megatron.core.fusions.fused_bias_geglu import quick_gelu, weighted_bias_qui
 from megatron.core.fusions.fused_bias_swiglu import weighted_bias_swiglu_impl
 from megatron.core.fusions.fused_weighted_squared_relu import weighted_squared_relu_impl
 from megatron.core.jit import jit_fuser
+from megatron.core.pipeline_parallel.fine_grained_activation_offload import (
+    fine_grained_offloading_group_commit,
+    fine_grained_offloading_group_start,
+    get_fine_grained_offloading_context,
+)
 from megatron.core.tensor_parallel.layers import (
     _initialize_affine_weight_cpu,
     _initialize_affine_weight_gpu,
@@ -34,6 +40,7 @@ from megatron.core.tensor_parallel.utils import divide
 from megatron.core.transformer.mlp import MLP, MLPSubmodules, apply_swiglu_sharded_factory
 from megatron.core.transformer.module import MegatronModule
 from megatron.core.transformer.moe import grouped_gemm_util as gg
+from megatron.core.transformer.moe.fused_fp8_padding import FusedFp8Padding
 from megatron.core.transformer.moe.moe_utils import ProcessGroupCollection
 from megatron.core.transformer.spec_utils import build_module
 from megatron.core.transformer.transformer_config import TransformerConfig
@@ -53,6 +60,23 @@ except ImportError:
 
     HAVE_TE = False
 
+try:
+    from aiak_fp8_quantizer import (
+        pad_and_dequant,
+        unpadding,
+    )
+    HAVE_FP8_QUANT = True
+except ImportError:
+    HAVE_FP8_QUANT = False
+
+logger = logging.getLogger(__name__)
+try:
+    from transformer_engine.pytorch.tensor.quantized_tensor import QuantizedTensor
+except Exception as e:
+    logger.warning(
+        f"QuantizedTensor import failed, FP8 not available: {str(e)}"
+    )
+
 
 # TODO(Hepteract): delete the usage of the global parallel_state.
 # Currently we still have to use the global parallel_state in expert_dist_ckpt_decorator(),
@@ -805,6 +829,16 @@ class TEGroupedMLP(MegatronModule):
             tp_group=pg_collection.expt_tp,
         )
 
+        self.offload_expert_fc1 = (
+            self.config.fine_grained_activation_offloading
+            and "expert_fc1" in self.config.offload_modules
+        )
+
+        self.offload_moe_act = (
+            self.config.fine_grained_activation_offloading
+            and "moe_act" in self.config.offload_modules
+        )
+
         self.activation_recompute = (
             self.config.recompute_granularity == 'selective'
             and "moe_act" in self.config.recompute_modules
@@ -814,10 +848,19 @@ class TEGroupedMLP(MegatronModule):
 
             set_save_original_input(self.linear_fc2)
 
+        # This is to avoid the CPU overhead of multiple d2h copies
+        if self.offload_expert_fc1 and not (self.config.fp8 or self.config.fp4):
+            from megatron.core.extensions.transformer_engine import set_save_original_input
+
+            set_save_original_input(self.linear_fc1)
+
         if self.config.fp8:
             assert HAVE_TE, "FP8 requires TE."
             self.fp8_padding = Fp8Padding(self.num_local_experts)
             self.fp8_unpadding = Fp8Unpadding(self.num_local_experts)
+            if HAVE_FP8_QUANT:
+                padding_align_size = get_fp8_align_size(self.config.fp8_recipe)
+                self.fused_fp8_padding = FusedFp8Padding(self.num_local_experts, padding_align_size)
 
     @staticmethod
     def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permuted_probs):
@@ -859,9 +902,14 @@ class TEGroupedMLP(MegatronModule):
         tokens_per_expert = tokens_per_expert.tolist()
         if self.config.fp8:
             actual_tokens_per_expert = tokens_per_expert
-            permuted_local_hidden_states, tokens_per_expert = self.fp8_padding(
-                permuted_local_hidden_states, tokens_per_expert
-            )
+            if isinstance(permuted_local_hidden_states, QuantizedTensor) and HAVE_FP8_QUANT:
+                permuted_local_hidden_states, tokens_per_expert = self.fused_fp8_padding(
+                    permuted_local_hidden_states, tokens_per_expert
+                )
+            else:
+                permuted_local_hidden_states, tokens_per_expert = self.fp8_padding(
+                    permuted_local_hidden_states, tokens_per_expert
+                )
             permuted_probs, _ = self.fp8_padding(
                 permuted_probs.unsqueeze(-1), actual_tokens_per_expert
             )
@@ -878,9 +926,21 @@ class TEGroupedMLP(MegatronModule):
             # Probs already applied, so reset to 1.
             permuted_probs = torch.ones_like(permuted_probs)
 
-        intermediate_parallel, bias_parallel = self.linear_fc1(
-            permuted_local_hidden_states, tokens_per_expert
-        )
+        if self.offload_expert_fc1:
+            permuted_local_hidden_states = fine_grained_offloading_group_start(
+                permuted_local_hidden_states, name="expert_fc1"
+            )
+        with get_fine_grained_offloading_context(self.offload_expert_fc1):
+            fc1_output, bias_parallel = self.linear_fc1(
+                permuted_local_hidden_states, tokens_per_expert
+            )
+        if self.offload_expert_fc1:
+            fc1_output, bias_parallel = fine_grained_offloading_group_commit(
+                fc1_output,
+                bias_parallel,
+                name="expert_fc1",
+                forced_released_tensors=[permuted_local_hidden_states],
+            )
 
         def bias_act_func(intermediate_parallel, bias_parallel, permuted_probs):
             if self.config.use_te_activation_func:
@@ -940,19 +1000,27 @@ class TEGroupedMLP(MegatronModule):
                 intermediate_parallel = intermediate_parallel.to(original_dtype)
             return intermediate_parallel
 
+        if self.offload_moe_act:
+            fc1_output = fine_grained_offloading_group_start(fc1_output, name="moe_act")
+
         if self.activation_recompute:
             self.activation_checkpoint = tensor_parallel.CheckpointWithoutOutput()
-            intermediate_parallel = self.activation_checkpoint.checkpoint(
-                bias_act_func, intermediate_parallel, bias_parallel, permuted_probs
-            )
-            output, output_bias = self.linear_fc2(intermediate_parallel, tokens_per_expert)
-            self.activation_checkpoint.discard_output_and_register_recompute(output)
+            with get_fine_grained_offloading_context(self.offload_moe_act):
+                bias_act_output = self.activation_checkpoint.checkpoint(
+                    bias_act_func, fc1_output, bias_parallel, permuted_probs
+                )
         else:
-            intermediate_parallel = bias_act_func(
-                intermediate_parallel, bias_parallel, permuted_probs
-            )
-            output, output_bias = self.linear_fc2(intermediate_parallel, tokens_per_expert)
+            with get_fine_grained_offloading_context(self.offload_moe_act):
+                bias_act_output = bias_act_func(fc1_output, bias_parallel, permuted_probs)
 
+        output, output_bias = self.linear_fc2(bias_act_output, tokens_per_expert)
+        if self.activation_recompute:
+            self.activation_checkpoint.discard_output_and_register_recompute(output)
+        if self.offload_moe_act:
+            (output,) = fine_grained_offloading_group_commit(
+                output, name="moe_act", forced_released_tensors=[fc1_output]
+            )
+            
         # upad and concat the output
         if self.config.fp8:
             output = self.fp8_unpadding(output, actual_tokens_per_expert)
