diff --git a/megatron/core/models/common/model_chunk_schedule_plan.py b/megatron/core/models/common/model_chunk_schedule_plan.py
index 6a411ccdc..ab66267bc 100644
--- a/megatron/core/models/common/model_chunk_schedule_plan.py
+++ b/megatron/core/models/common/model_chunk_schedule_plan.py
@@ -3,11 +3,15 @@
 from contextlib import nullcontext
 from typing import Optional
 
+import os
 import torch
 from torch import Tensor
 
 from megatron.core.enums import Fp8Recipe
 from megatron.core.fp8_utils import get_fp8_context
+from megatron.core.pipeline_parallel.fine_grained_activation_offload import (
+    fine_grained_offloading_set_last_layer,
+)
 from megatron.core.pipeline_parallel.utils import (
     AbstractSchedulePlan,
     NoopScheduleNode,
@@ -40,6 +44,7 @@ class TransformerLayerSchedulePlan:
     ├── moe_dispatch (TransformerLayerNode): dispatch All2All
     ├── mlp (TransformerLayerNode): mlp module
     ├── moe_combine (TransformerLayerNode): combine All2All
+    ├── post_combine (TransformerLayerNode): combine post process -> mlp_bda
     └── mtp_post_process (PostProcessNode): mtp post process
 
     Note that MTP layer has the same operation and execution order with TransformerLayer regarding
@@ -55,6 +60,7 @@ class TransformerLayerSchedulePlan:
     moe_dispatch = None
     mlp = None
     moe_combine = None
+    post_combine = None
     mtp_post_process = None
 
     def __init__(self, layer, event, chunk_state, comp_stream, comm_stream, extra_args={}):
@@ -133,6 +139,7 @@ class TransformerLayerSchedulePlan:
             moe_dispatch_module,
             mlp_module,
             moe_combine_module,
+            post_combine_module,
             mtp_post_process_module,
         ) = fwd_callables
 
@@ -144,10 +151,12 @@ class TransformerLayerSchedulePlan:
             self.post_attn = create_node(comp_stream, post_attn_module, "post_attn")
             self.moe_dispatch = create_node(comm_stream, moe_dispatch_module, "moe_dispatch")
             self.moe_combine = create_node(comm_stream, moe_combine_module, "moe_combine")
+            self.post_combine = create_node(comp_stream, post_combine_module, "post_combine")
         else:
             self.post_attn = NoopScheduleNode()
             self.moe_dispatch = NoopScheduleNode()
             self.moe_combine = NoopScheduleNode()
+            self.post_combine = NoopScheduleNode()
 
         if is_mtp:
             self.mtp_post_process = create_node(
@@ -170,19 +179,37 @@ class TransformerLayerSchedulePlan:
         )
 
     @staticmethod
-    def run(f_layer, b_layer, f_input=None, b_grad=None, is_last_layer_in_bwd=False):
+    def run(
+        f_layer, 
+        b_layer, 
+        f_input=None, 
+        b_grad=None, 
+        is_last_layer_in_bwd=False,
+        early_comm_launch=False,
+    ):
         """Schedule one-forward-one-backward operations for a single transformer layer.
 
         This function interleaves forward and backward operations, overlapping the communications
         (dispatch or combine) of one with the computations (att or mlp) of the other
         to maximize parallelism and efficiency.
 
+        Two execution modes are supported, controlled by CUDA_DEVICE_MAX_CONNECTIONS:
+
+        1) CUDA_DEVICE_MAX_CONNECTIONS = 1 (early_comm_launch=False):
+
         When f_layer and b_layer are not None, forward and backward pass are overlapped as follows:
         comm_stream: combine_bwd            | dispatch_fwd->dispatch_bwd  | combine_fwd
         comp_stream: attn_fwd->post_attn_fwd| mlp_bwd->mlp_bwd_dw->mlp_fwd| post_attn_bwd->attn_bwd
         For MTP, mtp_post_process_fwd is executed after the combine_fwd in the comp_stream,
         and mtp_post_process_bwd is executed before the combine_bwd in the comp_stream.
 
+        2) CUDA_DEVICE_MAX_CONNECTIONS = 32 (early_comm_launch = True):
+
+        In this mode, CUDA allows more concurrent kernel launches.
+        To better handle the operator sparsity at the beginning of the MLP backward phase,
+        the launch of MLP backward is moved ahead of the dispatch forward,
+        so as to improve overlap with forward computation and communication.
+
         Args:
             f_layer (TransformerLayerSchedulePlan): Forward layer (for current microbatch)
             b_layer (TransformerLayerSchedulePlan): Backward layer (for previous microbatch)
@@ -190,13 +217,15 @@ class TransformerLayerSchedulePlan:
             b_grad (Tensor): Gradient for backward computation
             is_last_layer_in_bwd (bool):
                 Whether the current layer is the last layer in the backward pass.
-
+            early_comm_launch (bool):
+                Enable fine-grained communication / computation overlap
         Returns:
             Functions or values for next iteration's computation
         """
 
         if b_layer is not None:
             b_grad = b_layer.mtp_post_process.backward(b_grad)
+            b_grad = b_layer.post_combine.backward(b_grad)
             b_grad = b_layer.moe_combine.backward(b_grad)
 
         if f_layer is not None:
@@ -204,15 +233,21 @@ class TransformerLayerSchedulePlan:
                 f_input = f_layer.attn.forward(f_input)
                 f_input = f_layer.post_attn.forward(f_input)
 
+        if early_comm_launch:
+            if f_layer is not None:
+                with f_layer.get_fp8_context():
+                    f_input = f_layer.moe_dispatch.forward(f_input)
+
         if b_layer is not None:
-            b_grad = b_layer.mlp.backward(b_grad)
+            with b_layer.get_fp8_context():
+                b_grad = b_layer.mlp.backward(b_grad)
 
-        if f_layer is not None:
-            with f_layer.get_fp8_context():
-                f_input = f_layer.moe_dispatch.forward(f_input)
+        if not early_comm_launch:
+            if f_layer is not None:
+                with f_layer.get_fp8_context():
+                    f_input = f_layer.moe_dispatch.forward(f_input)
 
         if b_layer is not None:
-            b_layer.mlp.backward_dw()
             b_grad = b_layer.moe_dispatch.backward(b_grad)
 
         if f_layer is not None:
@@ -222,11 +257,21 @@ class TransformerLayerSchedulePlan:
         if f_layer is not None:
             with f_layer.get_fp8_context():
                 f_input = f_layer.moe_combine.forward(f_input)
-                f_input = f_layer.mtp_post_process.forward(f_input)
 
         if b_layer is not None:
+            b_layer.mlp.backward_dw()
             b_grad = b_layer.post_attn.backward(b_grad)
-            b_grad = b_layer.attn.backward(b_grad)
+
+        if f_layer is not None:
+            with f_layer.get_fp8_context():
+                f_input = f_layer.post_combine.forward(f_input)
+                f_input = f_layer.mtp_post_process.forward(f_input)
+
+        # Delay the last attn_bwd in backward pass
+        # for overlapping with the p2p comm            
+        if b_layer is not None and not is_last_layer_in_bwd:
+            with b_layer.get_fp8_context():
+                b_grad = b_layer.attn.backward(b_grad)
 
         # Delay the last attn_dw in backward pass (attn_dw of the first layer)
         # for overlapping with the p2p comm
@@ -440,18 +485,22 @@ class TransformerModelChunkSchedulePlan(AbstractSchedulePlan):
         f_num_layers = f_schedule_plan.num_layers() if f_schedule_plan is not None else 0
         b_num_layers = b_schedule_plan.num_layers() if b_schedule_plan is not None else 0
         overlapped_layers = min(f_num_layers, b_num_layers)
+        early_comm_launch = os.environ.get("CUDA_DEVICE_MAX_CONNECTIONS") == "1"
 
         # combined forward and backward pass for overlapped layers
         for i in range(overlapped_layers):
             f_layer = f_schedule_plan.get_layer(i)
             b_layer = b_schedule_plan.get_layer(b_num_layers - 1 - i)
             torch.cuda.nvtx.range_push(f"layer_{i}f-layer_{b_num_layers - 1 - i}b")
+            if f_layer.layer.config.fine_grained_activation_offloading:
+                fine_grained_offloading_set_last_layer(i == f_num_layers - 1)
             f_input, b_grad = TransformerLayerSchedulePlan.run(
                 f_layer,
                 b_layer,
                 f_input=f_input,
                 b_grad=b_grad,
                 is_last_layer_in_bwd=(i == b_num_layers - 1),
+                early_comm_launch=early_comm_launch,
             )
             torch.cuda.nvtx.range_pop()
 
@@ -460,7 +509,11 @@ class TransformerModelChunkSchedulePlan(AbstractSchedulePlan):
             b_layer = b_schedule_plan.get_layer(b_num_layers - 1 - i)
             torch.cuda.nvtx.range_push(f"layer_{b_num_layers - 1 - i}b")
             _, b_grad = TransformerLayerSchedulePlan.run(
-                None, b_layer, b_grad=b_grad, is_last_layer_in_bwd=(i == b_num_layers - 1)
+                None,
+                b_layer,
+                b_grad=b_grad,
+                is_last_layer_in_bwd=(i == b_num_layers - 1),
+                early_comm_launch=early_comm_launch,
             )
             torch.cuda.nvtx.range_pop()
 
@@ -468,7 +521,14 @@ class TransformerModelChunkSchedulePlan(AbstractSchedulePlan):
         for i in range(overlapped_layers, f_num_layers):
             f_layer = f_schedule_plan.get_layer(i)
             torch.cuda.nvtx.range_push(f"layer_{i}f")
-            f_input, _ = TransformerLayerSchedulePlan.run(f_layer, None, f_input=f_input)
+            if f_layer.layer.config.fine_grained_activation_offloading:
+                fine_grained_offloading_set_last_layer(i == f_num_layers - 1)
+            f_input, _ = TransformerLayerSchedulePlan.run(
+                f_layer,
+                None,
+                f_input=f_input,
+                early_comm_launch=early_comm_launch,
+            )
             torch.cuda.nvtx.range_pop()
 
         if f_schedule_plan is not None and post_forward is not None:
@@ -478,6 +538,12 @@ class TransformerModelChunkSchedulePlan(AbstractSchedulePlan):
                 f_schedule_plan.wait_current_stream()
                 post_forward(f_input, f_schedule_plan.vp_stage)
 
+        # Delay the last attn_bwd in backward pass
+        # for overlapping with the p2p comm
+        if b_num_layers > 0:
+            with b_layer.get_fp8_context():
+                b_grad = b_schedule_plan.get_layer(0).attn.backward(b_grad)
+
         # post_backward()/send_backward_recv_backward() is running in the computation stream,
         # so the p2p comm could be overlapped with the wgrad of attn backward
         if b_schedule_plan is not None and post_backward is not None:
