diff --git a/xpu_plugin/_hydra/_linear_with_frozen_weight.py b/xpu_plugin/_hydra/_linear_with_frozen_weight.py
new file mode 100644
index 0000000..eb770b8
--- /dev/null
+++ b/xpu_plugin/_hydra/_linear_with_frozen_weight.py
@@ -0,0 +1,58 @@
+"""
+Reinplementation of `LinearWithFrozenWeight` via hydra.
+"""
+
+from typing import Optional
+
+import torch
+import torch.distributed as dist
+from hydrax.xaccelerator.linear import (
+    linear_bwd_dgrad,
+    linear_fwd,
+    linear_with_all_reduce,
+)
+from hydrax.xaccelerator.linear._linear import view_matrix
+from hydrax.xaccelerator.stateful import StatefulConfig
+from torch_xmlir.config import config
+
+
+@StatefulConfig(
+    retain_state_for_parameter=not config.disable_cast_cache
+).make_autograd_function_stateful
+class LinearWithFrozenWeight(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        input: torch.Tensor,
+        weight: torch.Tensor,
+        bias: Optional[torch.Tensor],
+        allreduce_dgrad: bool,
+        tp_group: Optional[dist.ProcessGroup],
+    ):
+        result = linear_fwd(input, weight, bias)
+
+        ctx.save_for_backward(weight)
+        ctx.allreduce_dgrad = allreduce_dgrad
+        ctx.tp_group = tp_group
+        ctx.input_shape = input.shape
+        ctx.input_dtype = input.dtype
+
+        return result
+
+    @staticmethod
+    def backward(ctx, grad_output: torch.Tensor):  # type: ignore[reportIncompatibleMethodOverride]
+        grad_input = grad_output.new_empty(ctx.input_shape, dtype=ctx.input_dtype)
+        (weight,) = ctx.saved_tensors
+
+        if ctx.allreduce_dgrad:
+            linear_with_all_reduce(
+                view_matrix(grad_output),
+                weight,
+                view_matrix(grad_input),
+                group=ctx.tp_group,
+                output=view_matrix(grad_input),
+            )
+        else:
+            linear_bwd_dgrad(grad_output, weight, grad_input)
+
+        return grad_input, None, None, None, None
