diff --git a/megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py b/megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py
index 868267584..526744e4e 100644
--- a/megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py
+++ b/megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py
@@ -130,7 +130,18 @@ class DistributedDataParallelConfig:
       This option will force to use conventional (local) userbuffer registration 
       when nccl_ub is set.
     """
+    use_precision_aware_optimizer: bool = False
+    """Enable the precision-aware optimizer in TransformerEngine.
+    This optimizer supports storing main parameters and optimizer states in reduced
+    precision formats such as FP16, BF16, or FP8.
 
+    When enabled in `DistributedDataParallelConfig`, Megatron-FSDP recognizes the
+    precision-aware optimizer and generates a `decoupled_grad`, a special gradient
+    that can have a different data type than the parameters.
+
+    Default is `False`.
+    """
+    
     def __post_init__(self):
         import os
 
