diff --git a/megatron/core/tensor_parallel/data.py b/megatron/core/tensor_parallel/data.py
index 6288e6af2..b157be083 100644
--- a/megatron/core/tensor_parallel/data.py
+++ b/megatron/core/tensor_parallel/data.py
@@ -4,7 +4,7 @@ import torch
 
 from megatron.core.utils import get_tensor_model_parallel_group_if_none
 
-_MAX_DATA_DIM = 5
+_MAX_DATA_DIM = 6
 
 
 def _check_data_types(keys, data, target_dtype):
