diff --git a/transformer_engine/common/fused_rope/fused_rope.cu b/transformer_engine/common/fused_rope/fused_rope.cu
index ccd0bc44..cfeca8f5 100644
--- a/transformer_engine/common/fused_rope/fused_rope.cu
+++ b/transformer_engine/common/fused_rope/fused_rope.cu
@@ -20,13 +20,15 @@ __device__ void fused_rope_block_forward(const scalar_t *src, const float *freqs
                                          const int offset_block, const int offset_block_dst,
                                          const int h, const int d, const int d2, const int stride_h,
                                          const int stride_d, const int o_stride_h,
-                                         const int o_stride_d) {
+                                         const int o_stride_d, const float mscale) {
   extern __shared__ float shared_mem_cos_sin[];
   float *shared_mem_cos = shared_mem_cos_sin;
   float *shared_mem_sin = shared_mem_cos_sin + d2;
   int tid = threadIdx.x * blockDim.y + threadIdx.y;
   for (int i = tid; i < d2; i += blockDim.x * blockDim.y) {
     sincosf(freqs[s_id * d2 + i], &shared_mem_sin[i], &shared_mem_cos[i]);
+    shared_mem_sin[i] *= mscale;
+    shared_mem_cos[i] *= mscale;
   }
   __syncthreads();
 
@@ -75,13 +77,16 @@ __device__ void fused_rope_block_backward(const scalar_t *src, const float *freq
                                           const int offset_block, const int offset_block_dst,
                                           const int h, const int d, const int d2,
                                           const int stride_h, const int stride_d,
-                                          const int o_stride_h, const int o_stride_d) {
+                                          const int o_stride_h, const int o_stride_d,
+                                          const float mscale) {
   extern __shared__ float shared_mem_cos_sin[];
   float *shared_mem_cos = shared_mem_cos_sin;
   float *shared_mem_sin = shared_mem_cos_sin + d2;
   int tid = threadIdx.x * blockDim.y + threadIdx.y;
   for (int i = tid; i < d2; i += blockDim.x * blockDim.y) {
     sincosf(freqs[s_id * d2 + i], &shared_mem_sin[i], &shared_mem_cos[i]);
+    shared_mem_sin[i] *= mscale;
+    shared_mem_cos[i] *= mscale;
   }
   __syncthreads();
 
@@ -137,7 +142,7 @@ __global__ void fused_rope_forward_kernel(const scalar_t *src, const int *cu_seq
                                           const int d2, const int stride_s_or_t, const int stride_b,
                                           const int stride_h, const int stride_d,
                                           const int o_stride_s_or_t, const int o_stride_b,
-                                          const int o_stride_h, const int o_stride_d) {
+                                          const int o_stride_h, const int o_stride_d, const float mscale) {
   int s_id = blockIdx.x, b_id = blockIdx.y;
   int offset_block, offset_block_dst;
   int cur_seqlens;
@@ -170,7 +175,7 @@ __global__ void fused_rope_forward_kernel(const scalar_t *src, const int *cu_seq
   }
 
   fused_rope_block_forward(src, freqs, dst, interleaved, s_id_for_freqs, offset_block,
-                           offset_block_dst, h, d, d2, stride_h, stride_d, o_stride_h, o_stride_d);
+                           offset_block_dst, h, d, d2, stride_h, stride_d, o_stride_h, o_stride_d, mscale);
 }
 
 template <typename scalar_t>
@@ -179,7 +184,7 @@ __global__ void fused_rope_backward_kernel(
     const bool interleaved, const int cp_size, const int cp_rank, const int s, const int h,
     const int d, const int d2, const int stride_s_or_t, const int stride_b, const int stride_h,
     const int stride_d, const int o_stride_s_or_t, const int o_stride_b, const int o_stride_h,
-    const int o_stride_d) {
+    const int o_stride_d, const float mscale) {
   int s_id = blockIdx.x, b_id = blockIdx.y;
   int offset_block, offset_block_dst;
   int cur_seqlens;
@@ -211,7 +216,7 @@ __global__ void fused_rope_backward_kernel(
   }
 
   fused_rope_block_backward(src, freqs, dst, interleaved, s_id_for_freqs, offset_block,
-                            offset_block_dst, h, d, d2, stride_h, stride_d, o_stride_h, o_stride_d);
+                            offset_block_dst, h, d, d2, stride_h, stride_d, o_stride_h, o_stride_d, mscale);
 }
 
 template <typename scalar_t>
@@ -465,7 +470,7 @@ void fused_rope_forward_launcher(const scalar_t *input, const int *cu_seqlens, c
                                  const NVTE_QKV_Format qkv_format, const bool interleaved,
                                  const int cp_size, const int cp_rank, const int s, const int b,
                                  const int h, const int d, const int d2, const int stride_s_or_t,
-                                 const int stride_b, const int stride_h, const int stride_d,
+                                 const int stride_b, const int stride_h, const int stride_d, const float mscale,
                                  cudaStream_t stream) {
   int warps_per_block = h < 16 ? 4 : 8;
   dim3 blocks(s, b);
@@ -489,7 +494,7 @@ void fused_rope_forward_launcher(const scalar_t *input, const int *cu_seqlens, c
   fused_rope_forward_kernel<<<blocks, threads, shared_mem_size, stream>>>(
       input, cu_seqlens, freqs, start_positions, output, interleaved, cp_size, cp_rank, s, h, d, d2,
       stride_s_or_t, stride_b, stride_h, stride_d, o_stride_s_or_t, o_stride_b, o_stride_h,
-      o_stride_d);
+      o_stride_d, mscale);
   NVTE_CHECK_CUDA(cudaGetLastError());
 }
 
@@ -499,7 +504,7 @@ void fused_rope_backward_launcher(const scalar_t *output_grads, const int *cu_se
                                   const NVTE_QKV_Format qkv_format, const bool interleaved,
                                   const int cp_size, const int cp_rank, const int s, const int b,
                                   const int h, const int d, const int d2, const int stride_s_or_t,
-                                  const int stride_b, const int stride_h, const int stride_d,
+                                  const int stride_b, const int stride_h, const int stride_d, const float mscale,
                                   cudaStream_t stream) {
   int warps_per_block = h < 16 ? 4 : 8;
   dim3 blocks(s, b);
@@ -523,7 +528,7 @@ void fused_rope_backward_launcher(const scalar_t *output_grads, const int *cu_se
   fused_rope_backward_kernel<<<blocks, threads, shared_mem_size, stream>>>(
       output_grads, cu_seqlens, freqs, input_grads, interleaved, cp_size, cp_rank, s, h, d, d2,
       stride_s_or_t, stride_b, stride_h, stride_d, o_stride_s_or_t, o_stride_b, o_stride_h,
-      o_stride_d);
+      o_stride_d, mscale);
   NVTE_CHECK_CUDA(cudaGetLastError());
 }
 
@@ -577,7 +582,7 @@ void fused_rope_forward(const Tensor &input, const Tensor &cu_seqlens, const Ten
                         const NVTE_QKV_Format qkv_format, const bool interleaved, const int cp_size,
                         const int cp_rank, const int s, const int b, const int h, const int d,
                         const int d2, const int stride_s_or_t, const int stride_b,
-                        const int stride_h, const int stride_d, cudaStream_t stream) {
+                        const int stride_h, const int stride_d, const float mscale, cudaStream_t stream) {
   TRANSFORMER_ENGINE_TYPE_SWITCH_INPUT(
       input.data.dtype, scalar_t,
       fused_rope_forward_launcher(reinterpret_cast<const scalar_t *>(input.data.dptr),
@@ -586,7 +591,7 @@ void fused_rope_forward(const Tensor &input, const Tensor &cu_seqlens, const Ten
                                   reinterpret_cast<const int *>(start_positions.data.dptr),
                                   reinterpret_cast<scalar_t *>(output->data.dptr), qkv_format,
                                   interleaved, cp_size, cp_rank, s, b, h, d, d2, stride_s_or_t,
-                                  stride_b, stride_h, stride_d, stream););
+                                  stride_b, stride_h, stride_d, mscale, stream););
 }
 
 void fused_rope_backward(const Tensor &output_grads, const Tensor &cu_seqlens, const Tensor &freqs,
@@ -594,7 +599,7 @@ void fused_rope_backward(const Tensor &output_grads, const Tensor &cu_seqlens, c
                          const bool interleaved, const int cp_size, const int cp_rank, const int s,
                          const int b, const int h, const int d, const int d2,
                          const int stride_s_or_t, const int stride_b, const int stride_h,
-                         const int stride_d, cudaStream_t stream) {
+                         const int stride_d, const float mscale, cudaStream_t stream) {
   TRANSFORMER_ENGINE_TYPE_SWITCH_INPUT(
       output_grads.data.dtype, scalar_t,
       fused_rope_backward_launcher(reinterpret_cast<const scalar_t *>(output_grads.data.dptr),
@@ -602,7 +607,7 @@ void fused_rope_backward(const Tensor &output_grads, const Tensor &cu_seqlens, c
                                    reinterpret_cast<const float *>(freqs.data.dptr),
                                    reinterpret_cast<scalar_t *>(input_grads->data.dptr), qkv_format,
                                    interleaved, cp_size, cp_rank, s, b, h, d, d2, stride_s_or_t,
-                                   stride_b, stride_h, stride_d, stream););
+                                   stride_b, stride_h, stride_d, mscale, stream););
 }
 
 void fused_qkv_rope_forward(const Tensor &qkv_input, const Tensor &q_freqs, const Tensor &k_freqs,
@@ -653,13 +658,13 @@ void nvte_fused_rope_forward(const NVTETensor input, const NVTETensor cu_seqlens
                              const bool interleaved, const int cp_size, const int cp_rank,
                              const int s, const int b, const int h, const int d, const int d2,
                              const int stride_s_or_t, const int stride_b, const int stride_h,
-                             const int stride_d, cudaStream_t stream) {
+                             const int stride_d, const float mscale, cudaStream_t stream) {
   NVTE_API_CALL(nvte_fused_rope_forward);
   using namespace transformer_engine;
   fused_rope_forward(*convertNVTETensorCheck(input), *convertNVTETensorCheck(cu_seqlens),
                      *convertNVTETensorCheck(freqs), *convertNVTETensorCheck(start_positions),
                      convertNVTETensorCheck(output), qkv_format, interleaved, cp_size, cp_rank, s,
-                     b, h, d, d2, stride_s_or_t, stride_b, stride_h, stride_d, stream);
+                     b, h, d, d2, stride_s_or_t, stride_b, stride_h, stride_d, mscale, stream);
 }
 
 void nvte_fused_rope_backward(const NVTETensor output_grads, const NVTETensor cu_seqlens,
@@ -668,13 +673,13 @@ void nvte_fused_rope_backward(const NVTETensor output_grads, const NVTETensor cu
                               const int cp_size, const int cp_rank, const int s, const int b,
                               const int h, const int d, const int d2, const int stride_s_or_t,
                               const int stride_b, const int stride_h, const int stride_d,
-                              cudaStream_t stream) {
+                              const float mscale, cudaStream_t stream) {
   NVTE_API_CALL(nvte_fused_rope_backward);
   using namespace transformer_engine;
   fused_rope_backward(*convertNVTETensorCheck(output_grads), *convertNVTETensorCheck(cu_seqlens),
                       *convertNVTETensorCheck(freqs), convertNVTETensorCheck(input_grads),
                       qkv_format, interleaved, cp_size, cp_rank, s, b, h, d, d2, stride_s_or_t,
-                      stride_b, stride_h, stride_d, stream);
+                      stride_b, stride_h, stride_d, mscale, stream);
 }
 
 void nvte_fused_qkv_rope_forward(const NVTETensor qkv_input, const NVTETensor q_freqs,
