diff --git a/transformer_engine/pytorch/csrc/extensions.h b/transformer_engine/pytorch/csrc/extensions.h
index 79fb7984..398f5174 100644
--- a/transformer_engine/pytorch/csrc/extensions.h
+++ b/transformer_engine/pytorch/csrc/extensions.h
@@ -343,12 +343,12 @@ at::Tensor fused_rope_forward(const at::Tensor &input, const at::Tensor &freqs,
                               const std::optional<at::Tensor> start_positions,
                               const NVTE_QKV_Format qkv_format, const bool interleaved,
                               const std::optional<at::Tensor> cu_seqlens, const int cp_size,
-                              const int cp_rank);
+                              const int cp_rank, const float mscale);
 
 at::Tensor fused_rope_backward(const at::Tensor &output_grads, const at::Tensor &freqs,
                                const NVTE_QKV_Format qkv_format, const bool interleaved,
                                const std::optional<at::Tensor> cu_seqlens, const int cp_size,
-                               const int cp_rank);
+                               const int cp_rank, const float mscale);
 
 std::tuple<at::Tensor, at::Tensor, at::Tensor> fused_qkv_rope_forward(
     const at::Tensor &qkv_input, const at::Tensor &q_freqs, const at::Tensor &k_freqs,
