diff --git a/diffsynth/trainers/unified_dataset.py b/diffsynth/trainers/unified_dataset.py
index c764ebd..445b7e1 100644
--- a/diffsynth/trainers/unified_dataset.py
+++ b/diffsynth/trainers/unified_dataset.py
@@ -312,6 +312,7 @@ class UnifiedDataset(torch.utils.data.Dataset):
             data = self.data[data_id % len(self.data)].copy()
             for key in self.data_file_keys:
                 if key in data:
+                    data["video_name"] = data["video"]
                     if key in self.special_operator_map:
                         data[key] = self.special_operator_map[key]
                     elif key in self.data_file_keys:
diff --git a/diffsynth/trainers/utils.py b/diffsynth/trainers/utils.py
index 0711176..5e84688 100644
--- a/diffsynth/trainers/utils.py
+++ b/diffsynth/trainers/utils.py
@@ -577,14 +577,19 @@ def launch_data_process_task(
     dataloader = torch.utils.data.DataLoader(dataset, shuffle=False, collate_fn=lambda x: x[0], num_workers=num_workers)
     accelerator = Accelerator()
     model, dataloader = accelerator.prepare(model, dataloader)
-
+    if accelerator.process_index == 0:
+        print(f"The size of the dataset is {len(dataset)}")
+        print(f"The number of workers is {accelerator.num_processes}")
+
     for data_id, data in tqdm(enumerate(dataloader)):
         with accelerator.accumulate(model):
+            data_id = data["video_name"]
             with torch.no_grad():
-                folder = os.path.join(model_logger.output_path, str(accelerator.process_index))
-                os.makedirs(folder, exist_ok=True)
-                save_path = os.path.join(model_logger.output_path, str(accelerator.process_index), f"{data_id}.pth")
+                os.makedirs(model_logger.output_path, exist_ok=True)
+                save_path = os.path.join(model_logger.output_path, f"{data_id}.tensors.pth")
                 data = model(data, return_inputs=True)
+                model_input_keys = ["input_latents", "context", "y"]
+                data = {key: data[key] for key in model_input_keys if key in data}
                 torch.save(data, save_path)


@@ -619,6 +624,7 @@ def wan_parser():
     parser.add_argument("--save_steps", type=int, default=None, help="Number of checkpoint saving invervals. If None, checkpoints will be saved every epoch.")
     parser.add_argument("--dataset_num_workers", type=int, default=0, help="Number of workers for data loading.")
     parser.add_argument("--weight_decay", type=float, default=0.01, help="Weight decay.")
+    parser.add_argument("--task", type=str, default="train", help="train or data_process.")
     return parser


diff --git a/examples/wanvideo/model_training/full/Wan2.2-I2V-A14B-DataProcess.sh b/examples/wanvideo/model_training/full/Wan2.2-I2V-A14B-DataProcess.sh
new file mode 100644
index 0000000..36a3e16
--- /dev/null
+++ b/examples/wanvideo/model_training/full/Wan2.2-I2V-A14B-DataProcess.sh
@@ -0,0 +1,29 @@
+accelerate launch examples/wanvideo/model_training/train.py \
+  --dataset_base_path /mnt/cluster/data/example_video_dataset \
+  --dataset_metadata_path /mnt/cluster/data/example_video_dataset/metadata.csv \
+  --task "data_process" \
+  --height 480 \
+  --width 832 \
+  --num_frames 49 \
+  --dataset_repeat 1 \
+  --model_paths '[
+    [
+        "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/high_noise_model/diffusion_pytorch_model-00001-of-00006.safetensors",
+        "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/high_noise_model/diffusion_pytorch_model-00002-of-00006.safetensors",
+        "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/high_noise_model/diffusion_pytorch_model-00003-of-00006.safetensors",
+        "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/high_noise_model/diffusion_pytorch_model-00004-of-00006.safetensors",
+        "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/high_noise_model/diffusion_pytorch_model-00005-of-00006.safetensors",
+        "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/high_noise_model/diffusion_pytorch_model-00006-of-00006.safetensors"
+    ],
+    "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/models_t5_umt5-xxl-enc-bf16.pth",
+    "/mnt/cluster/models/Wan-AI/Wan2.2-I2V-A14B/Wan2.1_VAE.pth"
+]' \
+  --learning_rate 1e-5 \
+  --num_epochs 2 \
+  --remove_prefix_in_ckpt "pipe.dit." \
+  --output_path "./data_preprocess/Wan2.2-I2V-A14B_full" \
+  --trainable_models "dit" \
+  --extra_inputs "input_image" \
+  --use_gradient_checkpointing_offload \
+  --max_timestep_boundary 0.358 \
+  --min_timestep_boundary 0
\ No newline at end of file
diff --git a/examples/wanvideo/model_training/train.py b/examples/wanvideo/model_training/train.py
index 37494e7..29a1bdc 100644
--- a/examples/wanvideo/model_training/train.py
+++ b/examples/wanvideo/model_training/train.py
@@ -1,7 +1,7 @@
 import torch, os, json
 from diffsynth import load_state_dict
 from diffsynth.pipelines.wan_video_new import WanVideoPipeline, ModelConfig
-from diffsynth.trainers.utils import DiffusionTrainingModule, ModelLogger, launch_training_task, wan_parser
+from diffsynth.trainers.utils import DiffusionTrainingModule, ModelLogger, launch_training_task, wan_parser, launch_data_process_task
 from diffsynth.trainers.unified_dataset import UnifiedDataset
 os.environ["TOKENIZERS_PARALLELISM"] = "false"

@@ -82,8 +82,15 @@ class WanTrainingModule(DiffusionTrainingModule):
         return {**inputs_shared, **inputs_posi}


-    def forward(self, data, inputs=None):
-        if inputs is None: inputs = self.forward_preprocess(data)
+    def forward(self, data, inputs=None, return_inputs=False):
+        # Inputs
+        if inputs is None:
+            inputs = self.forward_preprocess(data)
+        else:
+            inputs = self.transfer_data_to_device(inputs, self.pipe.device)
+        if return_inputs: return inputs
+
+        # Loss
         models = {name: getattr(self.pipe, name) for name in self.pipe.in_iteration_models}
         loss = self.pipe.training_loss(**models, **inputs)
         return loss
@@ -126,4 +133,8 @@ if __name__ == "__main__":
         args.output_path,
         remove_prefix_in_ckpt=args.remove_prefix_in_ckpt
     )
-    launch_training_task(dataset, model, model_logger, args=args)
+    if args.task == "data_process":
+        launch_data_process_task(dataset, model, model_logger, args=args)
+        print(f"data tensor saved to {args.output_path}")
+    else:
+        launch_training_task(dataset, model, model_logger, args=args)
