diff --git a/megatron/core/pipeline_parallel/schedules.py b/megatron/core/pipeline_parallel/schedules.py
index e83f8d906..bc54c508a 100644
--- a/megatron/core/pipeline_parallel/schedules.py
+++ b/megatron/core/pipeline_parallel/schedules.py
@@ -9,7 +9,13 @@ from torch.autograd.variable import Variable
 
 from megatron.core import parallel_state
 from megatron.core.enums import ModelType
-from megatron.core.pipeline_parallel.p2p_communication import P2PCommunicator
+from megatron.core.pipeline_parallel.fine_grained_activation_offload import (
+    fine_grained_offloading_reset,
+)
+from megatron.core.pipeline_parallel.p2p_communication import (
+    P2PCommunicator,
+    p2p_comm_tensor_remove_padding,
+)
 from megatron.core.pipeline_parallel.utils import (
     is_pp_first_stage,
     is_pp_last_stage,
@@ -461,7 +467,7 @@ def backward_step(input_tensor, output_tensor, output_tensor_grad, model_type, c
     # will not participate in the computation.
     # This results in a tensor that does not require gradients.
     # In such cases, we intentionally skip the backward pass while preserving zero gradients.
-    if output_tensor[0].requires_grad:
+    if (output_tensor[0].requires_grad) and (output_tensor[0].grad_fn is not None):
         if config.deallocate_pipeline_outputs:
             custom_backward(output_tensor[0], output_tensor_grad[0])
         else:
@@ -562,6 +568,9 @@ def forward_backward_no_pipelining(
     if config.timers is not None:
         config.timers('forward-backward', log_level=1).start(barrier=config.barrier_with_L1_time)
 
+    if not forward_only and config.fine_grained_activation_offloading:
+        fine_grained_offloading_reset()
+
     no_sync_func = config.no_sync_func
     if no_sync_func is None:
         no_sync_func = contextlib.nullcontext
@@ -898,6 +907,9 @@ def forward_backward_pipelining_with_interleaving(
         adjust_tensor_shapes_fn is None
     ), "adjust_tensor_shapes_fn is not supported for interleaved pipeline parallelism"
 
+    if not forward_only and config.fine_grained_activation_offloading:
+        fine_grained_offloading_reset()
+
     if config.overlap_p2p_comm and config.batch_p2p_comm:
         raise ValueError("Can not use both overlap_p2p_comm and batch_p2p_comm")
 
@@ -1180,6 +1192,12 @@ def forward_backward_pipelining_with_interleaving(
         offset = num_released_microbatches(virtual_microbatch_id, model_chunk_id)
         input_tensor = input_tensors[model_chunk_id][microbatch_id - offset]
 
+        # add for config.reduce_variable_seq_shape_p2p_comm, remove padding.
+        if config.reduce_variable_seq_shape_p2p_comm:
+            input_tensor = p2p_comm_tensor_remove_padding(input_tensor,
+                                                          config.p2p_comm_fixed_seq_lengths_per_rank)
+            input_tensors[model_chunk_id][microbatch_id - offset] = input_tensor
+
         return input_tensor
 
     def forward_step_helper_postprocess(model_chunk_id, output_tensor, num_tokens):
@@ -1248,6 +1266,11 @@ def forward_backward_pipelining_with_interleaving(
         output_tensor = output_tensors[model_chunk_id].pop(0)
         output_tensor_grad = output_tensor_grads[model_chunk_id].pop(0)
 
+        # add for reduce_variable_seq_shape_p2p_comm, remove padding.
+        if config.reduce_variable_seq_shape_p2p_comm:
+            output_tensor_grad = p2p_comm_tensor_remove_padding(output_tensor_grad,
+                                                                config.p2p_comm_fixed_seq_lengths_per_rank)
+
         return input_tensor, output_tensor, output_tensor_grad
 
     def backward_step_helper_postprocess(virtual_microbatch_id):
@@ -2043,6 +2066,9 @@ def forward_backward_pipelining_without_interleaving(
     if config.timers is not None:
         config.timers('forward-backward', log_level=1).start(barrier=config.barrier_with_L1_time)
 
+    if not forward_only and config.fine_grained_activation_offloading:
+        fine_grained_offloading_reset()
+
     # Disable async grad reductions
     no_sync_func = config.no_sync_func
     if no_sync_func is None:
