diff --git a/megatron/core/transformer/attention.py b/megatron/core/transformer/attention.py
index 659402a39..a86f77180 100644
--- a/megatron/core/transformer/attention.py
+++ b/megatron/core/transformer/attention.py
@@ -22,6 +22,11 @@ from megatron.core.parallel_state import (
     get_tensor_model_parallel_rank,
     get_tensor_model_parallel_world_size,
 )
+from megatron.core.pipeline_parallel.fine_grained_activation_offload import (
+    fine_grained_offloading_group_commit,
+    fine_grained_offloading_group_start,
+    get_fine_grained_offloading_context,
+)
 from megatron.core.process_groups_config import ProcessGroupCollection
 from megatron.core.transformer.identity_op import IdentityOp
 from megatron.core.transformer.module import MegatronModule
@@ -106,6 +111,7 @@ class SelfAttentionSubmodules:
     linear_proj: Union[ModuleSpec, type] = None
     q_layernorm: Union[ModuleSpec, type] = None
     k_layernorm: Union[ModuleSpec, type] = None
+    apply_rotary_fn: Union[ModuleSpec, type] = None
 
 
 @dataclass
@@ -118,6 +124,7 @@ class CrossAttentionSubmodules:
     linear_kv: Union[ModuleSpec, type] = None
     core_attention: Union[ModuleSpec, type] = None
     linear_proj: Union[ModuleSpec, type] = None
+    apply_rotary_fn: Union[ModuleSpec, type] = None
 
 
 class Attention(MegatronModule, ABC):
@@ -187,6 +194,21 @@ class Attention(MegatronModule, ABC):
             self.config.recompute_granularity == 'selective'
             and "core_attn" in self.config.recompute_modules
         )
+        
+        self.offload_qkv_linear = (
+            self.config.fine_grained_activation_offloading
+            and "qkv_linear" in self.config.offload_modules
+        )
+
+        self.offload_core_attention = (
+            self.config.fine_grained_activation_offloading
+            and "core_attn" in self.config.offload_modules
+        )
+
+        self.offload_attn_proj = (
+            self.config.fine_grained_activation_offloading
+            and "attn_proj" in self.config.offload_modules
+        )
 
         # Output.
         self.linear_proj = build_module(
@@ -203,6 +225,11 @@ class Attention(MegatronModule, ABC):
             tp_group=self.pg_collection.tp,
         )
 
+        if hasattr(submodules, "apply_rotary_fn") and submodules.apply_rotary_fn is not None:
+            self.apply_rotary_fn = build_module(submodules.apply_rotary_fn)
+        else:
+            self.apply_rotary_fn = apply_rotary_pos_emb
+
         if (
             HAVE_TE
             and self.config.fp8
@@ -451,7 +478,7 @@ class Attention(MegatronModule, ABC):
         rotary_cos: Tensor,
         rotary_sin: Tensor,
         rotary_interleaved: bool = False,
-    ) -> (Tensor, Tensor):
+    ) -> Tuple[Tensor, Tensor]:
         """
         The flash decoding kernel will do the following in a single execution:
         1. Compute RoPE embedding with precomputed cos & sin tensors
@@ -728,6 +755,17 @@ class Attention(MegatronModule, ABC):
         qkv_output = self.get_query_key_value_tensors(
             hidden_states, key_value_states, split_qkv=split_qkv
         )
+        if self.offload_qkv_linear:
+            hidden_states = fine_grained_offloading_group_start(hidden_states, name="qkv_linear")
+        with get_fine_grained_offloading_context(self.offload_qkv_linear):
+            qkv_output = self.get_query_key_value_tensors(
+                hidden_states, key_value_states, split_qkv=split_qkv
+            )
+        if self.offload_qkv_linear:
+            qkv_output, _ = fine_grained_offloading_group_commit(
+                qkv_output, name="qkv_linear", forced_released_tensors=[hidden_states]
+            )
+
         attn_mask_type = self.attn_mask_type
         block_table = None
         if split_qkv:
@@ -822,10 +860,11 @@ class Attention(MegatronModule, ABC):
                 cu_seqlens_q = cu_seqlens_kv = None
 
             if split_qkv:
+                assert self.apply_rotary_fn is not None, "apply_rotary_fn must be defined"
                 if q_pos_emb is not None:
                     # TODO VIJAY: simplify
                     if inference_context is None or inference_context.is_static_batching():
-                        query = apply_rotary_pos_emb(
+                        query = self.apply_rotary_fn(
                             query,
                             q_pos_emb,
                             config=self.config,
@@ -838,7 +877,7 @@ class Attention(MegatronModule, ABC):
                             query, q_pos_emb, self.config, cu_seqlens_q, self.pg_collection.cp
                         )
                 if k_pos_emb is not None:
-                    key = apply_rotary_pos_emb(
+                    key = self.apply_rotary_fn(
                         key,
                         k_pos_emb,
                         config=self.config,
@@ -873,17 +912,20 @@ class Attention(MegatronModule, ABC):
                 packed_seq_params=packed_seq_params,
             )
         else:
+            if self.offload_core_attention and self.training:
+                query = fine_grained_offloading_group_start(query, name="core_attn")
             if inference_context is None or inference_context.is_static_batching():
                 # Static batching attention kernel.
-                core_attn_out = self.core_attention(
-                    query,
-                    key,
-                    value,
-                    attention_mask,
-                    attn_mask_type=attn_mask_type,
-                    attention_bias=attention_bias,
-                    packed_seq_params=packed_seq_params,
-                )
+                with get_fine_grained_offloading_context(self.offload_core_attention):
+                    core_attn_out = self.core_attention(
+                        query,
+                        key,
+                        value,
+                        attention_mask,
+                        attn_mask_type=attn_mask_type,
+                        attention_bias=attention_bias,
+                        packed_seq_params=packed_seq_params,
+                    )
 
             else:
                 # Dynamic batching attention kernel.
@@ -903,6 +945,10 @@ class Attention(MegatronModule, ABC):
                     block_table,
                 )
                 core_attn_out = rearrange(core_attn_out, 's b h d -> s b (h d)')
+            if self.offload_core_attention and self.training:
+                (core_attn_out,) = fine_grained_offloading_group_commit(
+                    core_attn_out, name="core_attn", forced_released_tensors=[query, key, value]
+                )
 
         if packed_seq_params is not None and packed_seq_params.qkv_format == 'thd':
             # reshape to same output shape as unpacked case
@@ -917,7 +963,14 @@ class Attention(MegatronModule, ABC):
         # =================
 
         nvtx_range_push(suffix="linear_proj")
-        output, bias = self.linear_proj(core_attn_out)
+        if self.offload_attn_proj:
+            core_attn_out = fine_grained_offloading_group_start(core_attn_out, name="attn_proj")
+        with get_fine_grained_offloading_context(self.offload_attn_proj):
+            output, bias = self.linear_proj(core_attn_out)
+        if self.offload_attn_proj:
+            output, bias = fine_grained_offloading_group_commit(
+                output, bias, name="attn_proj", forced_released_tensors=[core_attn_out]
+            )
         nvtx_range_pop(suffix="linear_proj")
 
         return output, bias
