diff --git a/megatron/core/transformer/module.py b/megatron/core/transformer/module.py
index 4fdcacb79..917ff840d 100644
--- a/megatron/core/transformer/module.py
+++ b/megatron/core/transformer/module.py
@@ -1,8 +1,9 @@
 # Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
 
 """Megatron Module."""
-from typing import Optional, Tuple
+from typing import Callable, Optional, Tuple
 
+import logging
 import torch
 from torch.autograd import Variable
 from torch.nn.parameter import Parameter
@@ -19,6 +20,8 @@ _FLOAT_TYPES = (torch.FloatTensor, torch.cuda.FloatTensor)
 _HALF_TYPES = (torch.HalfTensor, torch.cuda.HalfTensor)
 _BF16_TYPES = (torch.BFloat16Tensor, torch.cuda.BFloat16Tensor)
 
+logger = logging.getLogger(__name__)
+
 
 def param_is_not_shared(param):  # pylint: disable=missing-function-docstring
     return not hasattr(param, 'shared') or not param.shared
@@ -395,9 +398,10 @@ class Float16Module(MegatronModule):
 
         else:
             raise Exception('Either config.fp16 or config.bf16 should be True.')
-
+        
         self.float16_convertor = float16_convertor
 
+
     def set_input_tensor(self, input_tensor):  # pylint: disable=missing-function-docstring
         return self.module.set_input_tensor(input_tensor)
 
@@ -451,3 +455,35 @@ class Float16Module(MegatronModule):
         self, state_dict, strict=True
     ):  # pylint: disable=missing-function-docstring
         self.module.load_state_dict(state_dict, strict=strict)
+
+def restore_fp16module_inputs_to_fp32(
+    instance: torch.nn.Module,
+    config: TransformerConfig,
+    weight_is_fp32: Callable[[torch.nn.Module], bool] = None,
+    keep_fp32_outputs: bool = False
+    ):
+    """
+    convert inputs to fp32.
+    """
+    if weight_is_fp32 is None:
+        return
+
+    if config.fp16:
+        def float16_convertor(val):
+            return val.half()
+    elif config.bf16:
+        def float16_convertor(val):
+            return val.bfloat16()
+
+    if config.fp16 or config.bf16:
+        def forward_pre_hook_fn(module, inputs):
+            if weight_is_fp32(module):
+                return float16_to_fp32(inputs)
+
+        def forward_post_hook_fn(module, inputs, output):
+            return fp32_to_float16(output, float16_convertor)
+
+        instance.register_forward_pre_hook(forward_pre_hook_fn)
+        if not keep_fp32_outputs:
+            instance.register_forward_hook(forward_post_hook_fn)
+
