diff --git a/megatron/core/tensor_parallel/layers.py b/megatron/core/tensor_parallel/layers.py
index e6e65425b..6b4f83660 100644
--- a/megatron/core/tensor_parallel/layers.py
+++ b/megatron/core/tensor_parallel/layers.py
@@ -980,6 +980,9 @@ class ColumnParallelLinear(torch.nn.Module):
                     f"not {expected_shape} as expected"
                 )
 
+        if weight.dtype == torch.float32:
+            input_ = input_.float()
+
         bias = self.bias if not self.skip_bias_add else None
 
         if (
