diff --git a/megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py b/megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py
index 6f9999f08..3cd16e6c4 100644
--- a/megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py
+++ b/megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py
@@ -1,6 +1,7 @@
 # Copyright (c) 2025, NVIDIA CORPORATION and Alibaba PAI. All rights reserved.
 from collections import defaultdict
 from typing import Dict
+from ..muon import Muon
 
 import torch
 
@@ -171,7 +172,10 @@ class HybridDeviceOptimizer(torch.optim.Optimizer):
             d2h_event = self._cpu_optimizer_map_data_event.pop(cpu_optimizer, None)
             if d2h_event is not None:
                 d2h_event.synchronize()
-            cpu_optimizer.step(closure)
+            if isinstance(cpu_optimizer, Muon):
+                cpu_optimizer.step(self.cpu_copys_map_gpu_param)
+            else:
+                cpu_optimizer.step(closure)
 
         # Sync state and param_groups to HDO after each step.
         # NOTE: It is possible for the optimizer to change the properties
@@ -274,8 +278,10 @@ class HybridDeviceOptimizer(torch.optim.Optimizer):
                     param = param.detach().clone().cpu().pin_memory()
                     offload_params_numel += param.numel()
                     cpu_copy = True
-                if self.param_update_in_fp32 and param.dtype != torch.float32:
-                    param = param.detach().clone().float()
+                if self.param_update_in_fp32:
+                    # In FP8 case, the passed in param_groups is fp32 shard main param, so just do a self-reference
+                    if param.dtype != torch.float32:
+                        param = param.detach().clone().float()
                     param_to_fp32_param[orig_param] = param
 
                 if cpu_copy:
