diff --git a/xpu_plugin/__init__.py b/xpu_plugin/__init__.py
new file mode 100644
index 0000000..814fa84
--- /dev/null
+++ b/xpu_plugin/__init__.py
@@ -0,0 +1,165 @@
+"""
+Copyright (c) 2014 Baidu.com, Inc. All Rights Reserved
+This module provides megatron-core-aiak plugin.
+"""
+
+import importlib
+import os
+import sys
+
+from . import (
+    _hydra,
+    empty_package,
+    mock_megatron,
+    scaled_masked_softmax,
+    scaled_softmax,
+    scaled_upper_triang_masked_softmax,
+    multi_latent_attention,
+)
+
+__all__ = [
+    "empty_package",
+    "scaled_upper_triang_masked_softmax",
+    "scaled_masked_softmax",
+    "scaled_softmax",
+    "multi_latent_attention",
+]
+
+
+import_module_list = [
+    "megatron",
+    "megatron.training",
+    "megatron.core.fusions",
+    "megatron.core.transformer.moe.token_dispatcher",
+    "megatron.core.transformer.moe.moe_layer",
+    "megatron.core.transformer.multi_latent_attention",
+]
+
+
+def import_modules(verbose=True):
+    """
+    import_modules
+    """
+    for m in import_module_list:
+        try:
+            importlib.import_module(m)
+        except Exception as e:
+            if verbose:
+                print(f'[WARN] Import Module {m} failed: {e}')
+
+
+def replace_module(source: str, target: str, verbose=True):
+    """
+    replace_module
+    """
+    try:
+        dst_module = importlib.import_module(target)
+        sys.modules[source] = dst_module
+        if verbose:
+            print(f'[INFO] Module {source} replaced with {target}')
+    except Exception as e:
+        if verbose:
+            print(f'[WARN] Module {source} replaced with {target} failed: {e}')
+
+
+def replace_symbol(symbol_path, target, record_origin=True, verbose=True):
+    """
+    replace_symbol
+    """
+    parts = symbol_path.split('.')
+    module_name = parts[0]
+
+    if module_name not in sys.modules:
+        try:
+            importlib.import_module(module_name)
+        except Exception as e:
+            if verbose:
+                print(f'[WARN] Failed to import {module_name}: {e}')
+            return
+
+    obj = sys.modules[module_name]
+    try:
+        for part in parts[1:-1]:
+            obj = getattr(obj, part)
+        origin = getattr(obj, parts[-1])
+        setattr(obj, parts[-1], target)
+        if record_origin:
+            setattr(obj, f'__origin_{parts[-1]}', origin)
+        if verbose:
+            print(f'[INFO] Symbol {symbol_path} replaced with {target.__name__}')
+    except Exception as e:
+        if verbose:
+            print(f'[WARN] Symbol {symbol_path} replaced with {target.__name__} failed: {e}')
+
+
+def init_megatron_core_aiak_plugin():
+    """
+    init_megatron_core_aiak_plugin
+    """
+    import_modules()
+
+    # replace_module
+    replace_module("fused_layer_norm_cuda", "xpu_plugin.empty_package")
+    replace_module(
+        "scaled_upper_triang_masked_softmax_cuda", "xpu_plugin.scaled_upper_triang_masked_softmax"
+    )
+    replace_module("scaled_masked_softmax_cuda", "xpu_plugin.scaled_masked_softmax")
+    replace_module("scaled_softmax_cuda", "xpu_plugin.scaled_softmax")
+    
+    # replace_symbol
+    replace_symbol(
+        "megatron.legacy.model.fused_layer_norm.MixedFusedLayerNorm",
+        mock_megatron.MockMixedFusedLayerNorm,
+    )
+    replace_symbol("megatron.legacy.model.LayerNorm", mock_megatron.MockMixedFusedLayerNorm)
+    replace_symbol("megatron.legacy.fused_kernels.load", empty_package.empty_function)
+    replace_symbol(
+        "megatron.core.transformer.mlp.bias_swiglu_impl", mock_megatron.mock_bias_swiglu_impl
+    )
+    replace_symbol(
+        "megatron.legacy.model.fused_bias_gelu.bias_gelu_impl", mock_megatron.mock_bias_gelu_impl
+    )
+    replace_symbol(
+        "megatron.core.tensor_parallel.mappings._gather_along_first_dim",
+        mock_megatron.mock_gather_along_first_dim,
+    )
+    replace_symbol(
+        "megatron.core.tensor_parallel.mappings._reduce_scatter_along_first_dim",
+        mock_megatron.mock_reduce_scatter_along_first_dim,
+    )
+    replace_symbol(
+        "megatron.legacy.model.fused_softmax.FusedScaleMaskSoftmax.is_kernel_available",
+        mock_megatron.mock_is_kernel_available,
+    )
+    replace_symbol("megatron.legacy.model.fused_bias_gelu.bias_gelu", mock_megatron.mock_bias_gelu)
+    replace_symbol(
+        "megatron.core.transformer.moe.token_dispatcher.unpermute", mock_megatron.mock_unpermute
+    )
+    replace_symbol(
+        "megatron.core.fp8_utils._modify_underlying_storage_impl", empty_package.empty_function
+    )
+    replace_symbol(
+        "megatron.core.fp8_utils._quantize_param_shard_impl", empty_package.empty_function
+    )
+    replace_symbol(
+        "megatron.core.transformer.multi_latent_attention.MLASelfAttention", 
+        multi_latent_attention.MLASelfAttention
+    )
+    if os.getenv("XMLIR_USE_ROPE_FUSION") in ("true", "1", "True"):
+        replace_symbol(
+            "megatron.core.models.common.embeddings.rope_utils.apply_rotary_emb_flash", None
+        )
+        replace_symbol(
+            "megatron.core.models.common.embeddings.rotary_pos_embedding._apply_rotary_pos_emb_bshd",
+            mock_megatron.mock_apply_rotary_pos_emb_bshd,
+        )
+
+    if os.getenv("XMLIR_MOCK_ASYNC_LINEAR") not in ("false", "0", "False"):
+        replace_symbol(
+            "megatron.core.tensor_parallel.layers.LinearWithGradAccumulationAndAsyncCommunication",
+            _hydra.LinearWithGradAccumulationAndAsyncCommunication,
+        )
+        replace_symbol(
+            "megatron.core.tensor_parallel.layers.LinearWithFrozenWeight",
+            _hydra.LinearWithFrozenWeight,
+        )
