diff --git a/megatron/core/transformer/moe/fused_a2a.py b/megatron/core/transformer/moe/fused_a2a.py
index 00a840f2b..35bce9dea 100644
--- a/megatron/core/transformer/moe/fused_a2a.py
+++ b/megatron/core/transformer/moe/fused_a2a.py
@@ -12,11 +12,41 @@ try:
 except ImportError:
     HAVE_DEEP_EP = False
 
+try:
+    from aiak_fp8_quantizer import (
+        pad_and_dequant,
+        unpadding,
+    )
+    HAVE_FUSED_FP8 = True
+except ImportError:
+    HAVE_FUSED_FP8 = False
+
 import torch
+import logging
+from transformer_engine.pytorch.constants import TE_DType
+logger = logging.getLogger(__name__)
+try:
+    from transformer_engine.pytorch.tensor.float8_blockwise_tensor import (
+        Float8BlockQuantizer,
+        Float8BlockwiseQTensor,
+    )
+except Exception as e:
+    logger.warning(
+        f"Float8BlockQuantizer, Float8BlockwiseQTensor import failed, FP8 not available: {str(e)}"
+    )
+    Float8BlockQuantizer, Float8BlockwiseQTensor = None, None
 
 _buffer = None
 
 
+def per_token_cast_back(x_fp8: torch.Tensor, x_scales: torch.Tensor):
+    """raw func for cast fp8 tensor to bf16
+    """
+    x_fp32 = x_fp8.to(torch.float32).view(x_fp8.size(0), -1, 128)
+    x_scales = x_scales.view(x_fp8.size(0), -1, 1)
+    return (x_fp32 * x_scales).view(x_fp8.shape).to(torch.bfloat16)
+
+
 def get_hidden_bytes(x: torch.Tensor) -> int:
     """Calculate the number of hidden bytes for a tensor.
 
@@ -78,6 +108,7 @@ class FusedDispatch(torch.autograd.Function):
         group,
         async_finish=False,
         allocate_on_comm_stream=False,
+        enable_fp8_comm=False,
     ):
         """Forward pass of fused dispatch."""
         previous_event = None
@@ -99,6 +130,20 @@ class FusedDispatch(torch.autograd.Function):
             allocate_on_comm_stream=allocate_on_comm_stream,
         )
 
+        tmp_x = x
+        if enable_fp8_comm:
+            # in fp8 communication, we do not need the amax_epsilon and force_pow_2_scales parameter
+            _x_quantizer = Float8BlockQuantizer(
+                fp8_dtype=TE_DType[torch.float8_e4m3fn],
+                rowwise=True,
+                columnwise=False,
+                amax_epsilon=0.0,
+                force_pow_2_scales=False,
+                block_scaling_dim=1,
+            )
+            x_fp8 = _x_quantizer(x)
+            tmp_x = (x_fp8._rowwise_data.view(torch.float8_e4m3fn), x_fp8._rowwise_scale_inv.T.contiguous())
+
         # Do MoE dispatch
         # NOTES: the CPU will wait for GPU's signal to arrive,
         # so this is not compatible with CUDA graph
@@ -110,7 +155,7 @@ class FusedDispatch(torch.autograd.Function):
             handle,
             after_event_overlap,
         ) = buffer.dispatch(
-            x,
+            tmp_x,
             topk_idx=token_indices,
             topk_weights=token_probs,  # DeepEP only supports float32 probs
             num_tokens_per_rank=num_tokens_per_rank,
@@ -133,6 +178,24 @@ class FusedDispatch(torch.autograd.Function):
         ctx.allocate_on_comm_stream = allocate_on_comm_stream
         tokens_per_expert = torch.tensor(num_recv_tokens_per_expert_list)
 
+        if enable_fp8_comm:
+            tmp_recv_x = Float8BlockwiseQTensor(
+                rowwise_data=recv_x[0],
+                rowwise_scale_inv=recv_x[1].T.contiguous(),
+                columnwise_data=None,
+                columnwise_scale_inv=None,
+                fp8_dtype=TE_DType[torch.float8_e4m3fn],
+                quantizer=_x_quantizer,
+                is_2D_scaled=False,
+                shape=recv_x[0].shape,
+                dtype=torch.bfloat16
+            )
+
+            if HAVE_FUSED_FP8:
+                recv_x = tmp_recv_x
+            else:
+                recv_x = tmp_recv_x.dequantize(dtype=torch.bfloat16)
+
         return (recv_x, recv_token_indices, recv_token_probs, tokens_per_expert, handle)
 
     @staticmethod
@@ -156,14 +219,14 @@ class FusedDispatch(torch.autograd.Function):
         # Make sure current stream is synchronized
         if ctx.async_finish:
             after_event.current_stream_wait()
-        return grad_x, None, grad_token_probs, None, None, None, None
+        return grad_x, None, grad_token_probs, None, None, None, None, None
 
 
 class FusedCombine(torch.autograd.Function):
     """Fused combine operation for MoE output combining computation and communication."""
 
     @staticmethod
-    def forward(ctx, x, group, handle, async_finish=False, allocate_on_comm_stream=False):
+    def forward(ctx, x, group, handle, async_finish=False, allocate_on_comm_stream=False, enable_fp8_comm=False):
         """Forward pass of fused combine."""
         previous_event = None
         if async_finish:
@@ -184,6 +247,7 @@ class FusedCombine(torch.autograd.Function):
         ctx.group = group
         ctx.async_finish = async_finish
         ctx.allocate_on_comm_stream = allocate_on_comm_stream
+        ctx.enable_fp8_comm = enable_fp8_comm
         return combined_x, None
 
     @staticmethod
@@ -193,6 +257,8 @@ class FusedCombine(torch.autograd.Function):
         if ctx.async_finish:
             previous_event = EventOverlap(EventHandle())
         buffer = get_buffer(ctx.group, get_hidden_bytes(grad_output))
+
+        # if backward enable fp8 comm, it will decrease end-to-end performance
         grad_x, _, _, _, _, after_event = buffer.dispatch(
             grad_output.contiguous(),
             handle=ctx.handle,
@@ -203,7 +269,7 @@ class FusedCombine(torch.autograd.Function):
         # Make sure current stream is synchronized
         if ctx.async_finish:
             after_event.current_stream_wait()
-        return grad_x, None, None, None, None
+        return grad_x, None, None, None, None, None
 
 
 if HAVE_DEEP_EP:
@@ -216,6 +282,7 @@ if HAVE_DEEP_EP:
         group,
         async_finish=False,
         allocate_on_comm_stream=False,
+        enable_fp8_comm=False,
     ):
         """Perform fused dispatch operation if deep_ep is available.
 
@@ -238,9 +305,10 @@ if HAVE_DEEP_EP:
             group,
             async_finish,
             allocate_on_comm_stream,
+            enable_fp8_comm,
         )
 
-    def fused_combine(x, group, handle, async_finish=False, allocate_on_comm_stream=False):
+    def fused_combine(x, group, handle, async_finish=False, allocate_on_comm_stream=False, enable_fp8_comm=False):
         """Perform fused combine operation if deep_ep is available.
 
         Args:
@@ -252,7 +320,7 @@ if HAVE_DEEP_EP:
         Returns:
             Result of FusedCombine
         """
-        return FusedCombine.apply(x, group, handle, async_finish, allocate_on_comm_stream)
+        return FusedCombine.apply(x, group, handle, async_finish, allocate_on_comm_stream, enable_fp8_comm)
 
     def set_deepep_num_sms(num_sms):
         """Sets the number of SMs to use for DeepEP"""
