diff --git a/megatron/core/transformer/moe/token_dispatcher.py b/megatron/core/transformer/moe/token_dispatcher.py
index 82fb7b005..1a6d282ec 100644
--- a/megatron/core/transformer/moe/token_dispatcher.py
+++ b/megatron/core/transformer/moe/token_dispatcher.py
@@ -933,6 +933,7 @@ class _DeepepManager(_DispatchManager):
         router_topk: int,
         num_experts: int,
         config: TransformerConfig,
+        enable_fp8_comm: bool = False,
     ):
         """
         Initialize the DeepEP dispatcher.
@@ -954,6 +955,7 @@ class _DeepepManager(_DispatchManager):
         self.router_dtype = config.moe_router_dtype
         self.capacity_factor = config.moe_expert_capacity_factor
         self.permute_fusion = config.moe_permute_fusion
+        self.enable_fp8_comm = enable_fp8_comm
 
         # Metadata
         self.token_indices: Optional[torch.Tensor] = None
@@ -1000,6 +1002,7 @@ class _DeepepManager(_DispatchManager):
                 self.group,
                 async_finish=async_finish,
                 allocate_on_comm_stream=allocate_on_comm_stream,
+                enable_fp8_comm=self.enable_fp8_comm,
             )
         )
         self.handle = handle
@@ -1060,6 +1063,7 @@ class _DeepepManager(_DispatchManager):
             self.handle,
             async_finish=async_finish,
             allocate_on_comm_stream=allocate_on_comm_stream,
+            enable_fp8_comm=self.enable_fp8_comm,
         )
         # Release the handle after combine operation
         self.handle = None
@@ -1168,12 +1172,15 @@ class MoEFlexTokenDispatcher(MoETokenDispatcher):
         assert (
             self.config.moe_pad_expert_input_to_capacity is False
         ), "Flex token dispatcher does not support --moe-pad-expert-input-to-capacity"
+
+        self.enable_fp8_comm = self.config.enable_fp8_comm
         self._comm_manager = _DeepepManager(
             group=self.tp_ep_group,
             num_local_experts=self.num_local_experts,
             router_topk=self.tp_size * self.config.moe_router_topk,
             num_experts=self.tp_size * self.config.num_moe_experts,
             config=self.config,
+            enable_fp8_comm=self.enable_fp8_comm,
         )
 
     def set_shared_experts(self, shared_experts):
