diff --git a/megatron/training/global_vars.py b/megatron/training/global_vars.py
index 17bda9cba..78cb5caa1 100644
--- a/megatron/training/global_vars.py
+++ b/megatron/training/global_vars.py
@@ -190,7 +190,11 @@ def _set_wandb_writer(args):
         if args.wandb_exp_name == '':
             raise ValueError("Please specify the wandb experiment name!")
 
-        import wandb
+        if os.environ.get('ENABLE_SWANLAB', '0') == "1":
+            import swanlab as wandb
+        else:
+            import wandb
+
         if args.wandb_save_dir:
             save_dir = args.wandb_save_dir
         else:
