diff --git a/megatron/training/training.py b/megatron/training/training.py
index b63636710..ec17e6a06 100644
--- a/megatron/training/training.py
+++ b/megatron/training/training.py
@@ -89,6 +89,7 @@ from megatron.legacy.data.data_samplers import build_pretraining_data_loader
 from megatron.core.optimizer_param_scheduler import OptimizerParamScheduler
 from megatron.core.transformer.moe import upcycling_utils
 from megatron.core.transformer.moe.moe_utils import track_moe_metrics
+from megatron.core.transformer.experimental_attention_variant.dsa import DSAIndexerLossLoggingHelper
 from megatron.core.transformer.multi_token_prediction import MTPLossLoggingHelper
 from megatron.core.parallel_state import (
     destroy_global_memory_buffer,
@@ -850,6 +851,16 @@ def update_train_iters(args):
     print_rank_0(f'setting training iterations to {args.train_iters}')
 
 
+def print_module_param_dtypes(module):
+    """Print parameter data types of a given PyTorch module."""
+    for name, param in module.named_parameters():
+        print_rank_0(f"param {name}: {param.data.dtype}")
+
+    if hasattr(module, "named_buffers"):
+        for name, buffer in module.named_buffers():
+            print_rank_0(f"buffer {name}: {buffer.dtype}")
+
+
 def get_model(model_provider_func, model_type=ModelType.encoder_or_decoder, wrap_with_ddp=True):
     """Build the model."""
     args = get_args()
@@ -920,11 +931,37 @@ def get_model(model_provider_func, model_type=ModelType.encoder_or_decoder, wrap
         for model_module in model:
             model_module.cuda(torch.cuda.current_device())
 
+    fp32_training_weights = None
     # Fp16 conversion.
     if args.fp16 or args.bf16:
+        param_pattern = args.use_fp32_dtype_for_param_pattern
+        if param_pattern and not isinstance(param_pattern, list):
+            param_pattern = [param_pattern]
+
         config = get_model_config(model[0])
         model = [Float16Module(config, model_module) for model_module in model]
 
+        
+        fp32_training_weights = param_pattern
+        #covert fp32
+        if fp32_training_weights:
+            for module in zip(model):
+                if not isinstance(module, list):
+                    module = module[0]
+                for name, buf in module.module.named_parameters():
+                    if any(fp32_weight in name for fp32_weight in fp32_training_weights):
+                        buf.data = buf.data.to(dtype=torch.float32)
+                        print(f'verl check update param precison {name}')
+
+                for name, buf in module.module.named_buffers():
+                    if any(fp32_weight in name for fp32_weight in fp32_training_weights):
+                        buf.data = buf.data.to(dtype=torch.float32)
+                        print(f'verl check update buffer precison {name}')
+
+        if param_pattern:
+            print_rank_0("> model param_dtypes:")
+            print_module_param_dtypes(model[0])
+        
     # Materialize tensors on meta device (GPU allocation) if not using FSDP2 and not using Megatron FSDP.
     if args.init_model_with_meta_device and not args.use_torch_fsdp2 and not args.use_megatron_fsdp:
         #for model_module in model:
@@ -974,6 +1011,8 @@ def get_model(model_provider_func, model_type=ModelType.encoder_or_decoder, wrap
             kwargs['average_in_collective'] = args.ddp_average_in_collective
             if args.use_megatron_fsdp and args.use_precision_aware_optimizer:
                 kwargs["preserve_fp32_weights"] = False
+
+            kwargs["force_turn_on_bucketing"] = args.force_turn_on_bucketing
             ddp_config = DistributedDataParallelConfig(**kwargs)
 
             # In the Megatron FSDP and DDP use path, we need to initialize the bucket size.
@@ -1566,6 +1605,16 @@ def training_log(
         MTPLossLoggingHelper.track_mtp_metrics(
             mtp_loss_scale, iteration, writer, wandb_writer, total_loss_dict
         )
+    # Track sparse attention indexer loss
+    if args.dsa_indexer_loss_coeff is not None and args.dsa_indexer_loss_coeff > 0:
+        indexer_loss_scale = 1 / get_num_microbatches()
+        DSAIndexerLossLoggingHelper.track_indexer_metrics(
+            loss_scale=indexer_loss_scale,
+            iteration=iteration,
+            writer=writer,
+            wandb_writer=wandb_writer,
+            total_loss_dict=total_loss_dict,
+        )
     if iteration % args.log_interval == 0:
         if args.record_memory_history and is_last_rank():
             snapshot = torch.cuda.memory._snapshot()
@@ -1728,7 +1777,8 @@ def save_checkpoint_and_time(
 
     # Stop timer to get accurate train interval time and exclude checkpointing duration
     timers('interval-time').stop()
-    energy_monitor.pause()
+    if args.log_energy and energy_monitor is not None:
+        energy_monitor.pause()
 
     # Extra barrier is added to make sure all ranks report the max time.
     timer_key = 'save-checkpoint-non-persistent' if non_persistent_ckpt else 'save-checkpoint'
@@ -1770,7 +1820,8 @@ def save_checkpoint_and_time(
         )
 
     # Recover timing
-    energy_monitor.resume()
+    if args.log_energy and energy_monitor is not None:
+        energy_monitor.resume()
     timers('interval-time', log_level=0).start(barrier=True)
 
 
@@ -2807,6 +2858,11 @@ def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider
     # Rely on distributed-aware core datasets, temporary
     is_distributed = getattr(build_train_valid_test_datasets_provider, "is_distributed", False)
 
+    if args.preprocess_data_on_cpu:
+        device = torch.device("cpu")
+    else:
+        device = torch.device("cuda")
+
     # Construct the data pipeline
     if is_distributed or mpu.get_tensor_model_parallel_rank() == 0:
 
@@ -2838,12 +2894,15 @@ def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider
         do_valid = valid_dataloaders is not None and (args.full_validation or args.eval_iters > 0)
         do_test = test_dataloader is not None and (args.full_validation or args.eval_iters > 0)
         flags = torch.tensor(
-            [int(do_train), int(do_valid), int(do_test)], dtype=torch.long, device='cuda'
+            [int(do_train), int(do_valid), int(do_test)], dtype=torch.long, device=device
         )
     else:
-        flags = torch.tensor([0, 0, 0], dtype=torch.long, device='cuda')
+        flags = torch.tensor([0, 0, 0], dtype=torch.long, device=device)
 
-    torch.distributed.broadcast(flags, 0)
+    if args.preprocess_data_on_cpu:
+        print("Dataset preprocessing does not require this step.")
+    else:
+        torch.distributed.broadcast(flags, 0)
 
     args.do_train = getattr(args, "do_train", False) or flags[0].item()
     args.do_valid = getattr(args, "do_valid", False) or flags[1].item()
