diff --git a/megatron/core/distributed/distributed_data_parallel.py b/megatron/core/distributed/distributed_data_parallel.py
index df5bccd71..af00bc834 100644
--- a/megatron/core/distributed/distributed_data_parallel.py
+++ b/megatron/core/distributed/distributed_data_parallel.py
@@ -115,12 +115,13 @@ class DistributedDataParallel(_BaseDataParallel):
         # disable_bucketing is True (e.g., we might not want to break up model parameters
         # into buckets for model chunks after the first in the interleaved schedule).
         self.bucket_size = self.ddp_config.bucket_size
-        if isinstance(self.pp_group, list):
-            pp_rank = self.pp_group[0].rank()
-        else:
-            pp_rank = self.pp_group.rank()
-        if disable_bucketing or pp_rank > 0:
-            self.bucket_size = None
+        if not self.ddp_config.force_turn_on_bucketing:
+            if isinstance(self.pp_group, list):
+                pp_rank = self.pp_group[0].rank()
+            else:
+                pp_rank = self.pp_group.rank()
+            if disable_bucketing or pp_rank > 0:
+                self.bucket_size = None
 
         self.param_to_bucket_group = {}
 
