diff --git a/megatron/core/distributed/distributed_data_parallel_config.py b/megatron/core/distributed/distributed_data_parallel_config.py
index 6c4cab7d5..c7f1e5fcf 100644
--- a/megatron/core/distributed/distributed_data_parallel_config.py
+++ b/megatron/core/distributed/distributed_data_parallel_config.py
@@ -135,6 +135,21 @@ class DistributedDataParallelConfig:
     delay_wgrad_compute: bool = False
     """Delay the weight gradient computation to improve batch-level communication overlapping"""
 
+    use_precision_aware_optimizer: bool = False
+    """Enable the precision-aware optimizer in TransformerEngine.
+    This optimizer supports storing main parameters and optimizer states in reduced
+    precision formats such as FP16, BF16, or FP8.
+
+    When enabled in `DistributedDataParallelConfig`, Megatron-FSDP recognizes the
+    precision-aware optimizer and generates a `decoupled_grad`, a special gradient
+    that can have a different data type than the parameters.
+
+    Default is `False`.
+    """
+
+    force_turn_on_bucketing: bool = False
+    """Force turn on parameter bucketing"""
+
     def __post_init__(self):
         import os
 
